---
output: 
#   word_document:
#     fig_caption: true
#     reference_docx: word-styles-reference.docx
# always_allow_html: true
  pdf_document:
    keep_tex: true
    fig_caption: true
    latex_engine: pdflatex
# see http://svmiller.com/blog/2016/02/svm-r-markdown-manuscript/ for details on this .Rmd template

title: "Epigenomic assessment of cardiovascular disease risk and interactions with traditional risk metrics"

# author: |
#   | Kenneth Westerman$^1$, Alba Fernández-Sanlés$^2,3^$, Prasad Patil$^4$, Paola Sebastiani$^4$, Paul Jacques$^1$, John M. Starr$^5,6^$, Ian Deary$^5,6^$, Qing Liu$^7$, Simin Liu$^7$, Roberto Elosua$^2,8,9^$, Dawn L. DeMeo$^10^$, José M. Ordovás$^1,11,12^$
#   | $^1$JM-USDA Human Nutrition Research Center on Aging at Tufts University, Boston, MA, USA, $^2$Cardiovascular Epidemiology and Genetics Research Group, REGICOR Study group, IMIM (Hospital del Mar Medical Research Institute), Barcelona, Catalonia, Spain, $^3^Pompeu Fabra University (UPF), Barcelona, Catalonia, Spain, $^4$Department of Biostatistics, Boston University School of Public Health, Boston, MA, USA, $^5$Department of Psychology, University of Edinburgh, Edinburgh, UK, $^6$Centre for Cognitive Ageing and Cognitive Epidemiology, University of Edinburgh, Edinburgh, UK, $^7$Department of Epidemiology, Brown University School of Public Health, Providence, RI, USA, $^8$ CIBER Cardiovascular Diseases (CIBERCV), Spain,$^9$ Medicine Department, Medical School, University of Vic-Central University of Catalonia (UVic-UCC), Vic, Catalonia, Spain,  $^10$Channing Division of Network Medicine, Department of Medicine, Brigham and Women's Hospital, Boston, MA, USA, $^11$IMDEA Alimentación, CEI, UAM, Madrid, Spain, $^12$Centro Nacional de Investigaciones Cardiovasculares (CNIC), Madrid, Spain

documentclass: article

bibliography: mrs_manuscript.bib
csl: jaha.csl

header-includes:
  - \usepackage{booktabs}
  - \usepackage{longtable}
  - \usepackage{array}
  - \usepackage{multirow}
  - \usepackage{wrapfig}
  - \usepackage{float}
  - \usepackage{pdflscape}
  - \usepackage{tabu}
  - \usepackage{threeparttable}
  - \usepackage{threeparttablex}
  - \usepackage[normalem]{ulem}
  - \usepackage{makecell}
---

```{r setup, include=F}
knitr::opts_chunk$set(echo=F, fig.path="figures/", 
                      cache.path="../../cache/mrs_manuscript/",
                      warning=F)
# options(kableExtra.auto_format=F)
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra",
    "tidyverse", "broom", "cowplot", 
    "survival", "survminer",
    "IlluminaHumanMethylation450kanno.ilmn12.hg19"), 
  library, character.only=T))
theme_set(theme_cowplot(font_size=8))
```

```{r load-results}
load("../../output/mrs_models_objects.RData")

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
anno_450k <- data.frame(
  getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19), 
  stringsAsFactors=F)

nmd_all_preimputation <- mutate(nmd_all_preimputation,
                                diabetes=dm_med | (glu > 125))
diabetes_tbl <- table(is.na(nmd_all_preimputation$diabetes), 
                      nmd_all_preimputation$batch)
ht_med_tbl <- table(is.na(nmd_all_preimputation$ht_med), 
                    nmd_all_preimputation$batch)
smk_now_tbl <- table(is.na(nmd_all_preimputation$smk_now), 
                     nmd_all_preimputation$batch)
get_missing_pct <- function(tbl) {
  frac <- tbl[2, "fhs_UofMinn"] / sum(tbl[, "fhs_UofMinn"])
  frac * 100
}
diabetes_missing_pct <- get_missing_pct(diabetes_tbl)
ht_med_missing_pct <- get_missing_pct(ht_med_tbl)
smk_now_missing_pct <- get_missing_pct(smk_now_tbl)
```

# Abstract

Background: Epigenome-wide association studies for cardiometabolic risk factors have discovered multiple loci associated with incident cardiovascular disease (CVD). However, few studies have sought to directly optimize a predictor of CVD risk. Furthermore, it is challenging to train multivariate models across multiple studies in the presence of study- or batch effects. 

Methods and Results: Here, we analyzed existing DNA methylation data collected using the Illumina HumanMethylation450 microarray to create a predictor of CVD risk across three cohorts: Women’s Health Initiative, Framingham Heart Study Offspring Cohort, and Lothian Birth Cohorts. We trained Cox proportional hazards-based elastic net regressions for incident CVD separately in each cohort, and used a recently-introduced cross-study learning approach to integrate these individual scores into an ensemble predictor. The methylation-based risk score (MRS) was associated with CVD time-to-event in a held-out fraction of the Framingham dataset (HR per SD=1.28, 95% CI: 1.10-1.50) and predicted myocardial infarction status in the independent REGICOR dataset (OR per SD=2.14, 95% CI: 1.58-2.89). These associations remained after adjustment for traditional cardiovascular risk factors and were similar to those from elastic net models trained on a directly merged dataset. Additionally, we investigated interactions between the MRS and both genetic and biochemical CVD risk, showing preliminary evidence of an enhanced performance in those with less traditional risk factor elevation.

Conclusions: This investigation provides proof-of-concept for a genome-wide, CVD-specific epigenomic risk score and suggests that the DNA methylation data may enable the discovery of high-risk individuals that would be missed by alternative risk metrics.

Key Words: Epigenomics, DNA methylation, cardiovascular disease, risk prediction

# Introduction

DNA methylation is an important epigenetic pathway through which genetic variants and environmental exposures impact disease risk [@Bonder2016; @Tobi2018]. Methylation at specific cytosine-phosphate-guanine (CpG) sites has been associated with disease in epigenome-wide association studies, even showing associations in blood as a convenient but non-target tissue such as for type 2 diabetes [@Bacos2016]. Methylation-based risk scores allow genome-wide aggregation of epigenetic information, similarly to the more established genetic risk scores, and allow for the use of models with arbitrary complexity. These risk scores are often developed initially by using methylation as a proxy for disease risk factors, such as body mass index (BMI) [@Wahl2017] and general aging-related morbidity [@Levine2018]. Alternatively, given sufficient sample size, epigenetic associations with disease risk can be modeled directly [@Hao2017].

Associations between DNA methylation and cardiovascular disease (CVD) have been explored in many different cohorts and using diverse approaches. Cross-sectional associations have been found across multiple relevant tissues, namely blood, aorta, and other vascular tissues [@Fernandez-Sanles2017]. Some investigations aimed at cardiovascular risk factors have discovered CpGs predictive of CVD development [@Hedman2017; @Aslibekyan2018], while Mendelian randomization approaches have suggested causality of at least some of these CpG-risk factor associations [@Richardson2017]. A few studies directly modeling incident CVD as a primary outcome have either been conducted using only global (not locus-specific) methylation levels [@Baccarelli2010], or have found limited additional predictive power in the presence of known risk factors [@Guarrera2015]. A recent large-scale meta-analysis found multiple CpG sites predictive of incident coronary heart disease, but focused on univariate approaches [@Agha2019]. We have previously investigated methylation regions and modules associating with incident CVD, generating mechanistic insights but without aggregating these results into a direct predictor of risk [@Westerman2019]. Additionally, it is unclear how the CVD risk tracked by DNA methylation is redundant with or complementary to existing risk metrics, including genetic scores [@Khera2018] and those based on traditional cardiovascular risk factors (e.g. the Framingham Risk Score for generalized CVD) [@DAgostino2008].

Combining signal across population-scale cohorts can increase sample size while attenuating the effect of study-specific biases and confounding factors, but can be prone to emergent sources of confounding from "batch" effects or other systematic biases in methylation data across cohorts. This is especially problematic when there is notable class imbalance (i.e. different outcome frequencies) across cohorts [@Goh2017]. The most common method for dealing with this heterogeneity is meta-analysis, but standard meta-analysis approaches are restricted to univariate (one CpG site at a time) models. Other approaches include batch effect correction on the input dataset (e.g. ComBat [@Johnson2007]), direct adjustment for batch/study in linear models, or adjustment for derived variables intended to capture technical biases (e.g. surrogate variable analysis [@Leek2007]), but these approaches can often lead to over- or under-estimates of true biological effects [@Goh2017]. An alternative approach described recently, cross-study learning, instead trains an ensemble predictor consisting of one or multiple models per cohort [@Patil2018]. This strategy allows the use of arbitrarily complex models while avoiding technical confounding from direct combination of the datasets.

In order to develop an improved DNA methylation-based cardiovascular risk predictor using multiple heterogeneous training cohorts, we used a cross-study learning method to develop an ensemble of penalized time-to-event regression risk models. The resulting composite risk score performed well in a held-out data subset, associating with survival even in the presence of traditional risk factors, and showing similar performance to models trained on naively merged datasets. External validation was achieved in a case-control study for prevalent myocardial infarction (MI). Further, interactions were assessed between the composite methylation-based risk score and other risk predictors, finding that it is potentially most effective in those with low Framingham Risk Scores.

# Methods

## Study participants and phenotype collection

Phenotypes (demographic, anthropometric, biochemical, and clinical), DNA methylation data, and imputed genotypes were available either from publicly available controlled-access databases or upon request from the cohorts. Cohort-specific details are provided in Supplementary methods. Blood-based biochemical markers (total cholesterol, LDL-cholesterol, HDL-cholesterol, triglycerides, fasting glucose, high-sensitivity C-reactive protein, and systolic blood pressure) were log10-transformed for all analyses. In the Lothian Birth Cohort 1936, LDL was estimated from total cholesterol and triglycerides using the Friedewald equation. Diabetes was defined as either use of diabetes medication or a measured fasting blood glucose level of >125 mg/dL. Antihypertensive medication use, smoking status, and diabetes status were assumed to be false where missing, though missing data rates for these variables in the held-out FHS subset were low (`r round(ht_med_missing_pct, 1)`%, `r round(smk_now_missing_pct, 1)`%, and `r round(diabetes_missing_pct)`%, respectively). Analysis of these datasets was approved by the Tufts University Health Sciences Institutional Review Board (protocol 12592), and all subjects gave informed consent. <!--While directly available in WHI, pack-years of smoking was approximated in FHS by multiplying either the number of years since starting smoking by the current number of packs per day, or else by multiplying the number of years smoked by the median number of packs per day (15) for those who had quit. Framingham Risk Scores were calculated as described previously [@DAgostino2008].-->

## DNA methylation data processing

DNA methylation data for all initial cohorts (WHI, FHS, and LBC) were collected using the Illumina HumanMethylation450 microarray platform [@Bibikova2011] and downloaded as raw intensity files. FHS methylation data were collected in two primary batches in two centers -- one in subjects from a nested case-control for CVD measured at Johns Hopkins University (FHS-JHU), and the other in a larger set of remaining Framingham Offspring participants measured at the University of Minnesota (FHS-UM). Preprocessing was performed using the *minfi* and *wateRmelon* packages for R [@Aryee2014; @Pidsley2013]. Sample-wise filters were as follows: robust overall signal in the main cluster based on visual inspection of an intensity plot, less than 10% of probes undetected at a detection threshold of p<1e-16, and a reported sex matching methylation-based sex prediction. Probes were removed using the following criteria: more than 10% of samples undetected at a detection threshold of p<1e-16, location in the X or Y chromosomes, non-CpG probes, cross-hybridizing probes, probes measuring SNPs, and probes with an annotated SNP at the CpG site or in the single-base extension region. Samples were normalized using the Noob method for background correction and dye-bias normalization, followed by the BMIQ method for probe type correction [@Fortin2016; @Teschendorff2013]. Blood cell fractions for 6 blood cell types (CD4+ T-cells, CD8+ T-cells, B-cells, natural killer cells, monocytes, and granulocytes) were estimated using a common reference-based method [@Houseman2012], and 5 of these (excluding granulocytes) were included in cell count-adjusted statistical models. After quality control and filtering steps, `r length(common_cpgs)` CpG sites were shared between the 3 datasets, formatted as beta values (roughly equal to the ratio of methylated signal to total microarray signal, or $\beta=\frac{M}{M+U+100}$).

DNA methylation data for the REGICOR cohort were collected using the Illumina MethylationEPIC microarray platform [@Pidsley2016] and analyzed using the *wateRmelon* [@Pidsley2013] and methylumi [@Davis2019] R packages. Samples were excluded based on detection p-value >0.05 in at least 1% of probes or failure to cluster in the appropriate sex based on X chromosome methylation. Probes were excluded based on detection p-value >0.05 in at least 1% of samples, a bead count <3 in at least 5% of samples, discarding by Illumina based on underperformance (n=1,031) or changes in the manufacturing process (n=977), non-CpG targets, and cross-hybridization (n=43,979). A batch normalization was performed by standardizing beta values to mean zero and unit variance within each bisulfite conversion batch prior to analysis. After quality control and preprocessing, 811,610 CpG sites across 391 individuals were available for analysis. Participants were further excluded from analysis due to unknown smoking habits (n=10) and unavailable information regarding diabetes, hypertension, or hyperlipidemia (n=53). Surrogate variable analysis [@Leek2007] was used to calculate two surrogate variables, representing potential technical and biological confounders, for adjustment in MRS replication models.

## CVD risk modeling

Study-specific CVD risk models were trained using penalized Cox proportional hazards regressions with the elastic net penalty. CVD events were defined as including coronary heart disease, stroke, and death from CVD (see Supplementary Methods for cohort-specific details), and times were right-censored based on the most recent exam available in each cohort. The elastic alpha parameter was initially set at 0.05 (closer to ridge regression) in order to retain a higher number of CpGs with non-zero weights while still performing feature selection [@Benton2017]. Inner cross-validation loops varying alpha between 0.05 and 0.95 showed negligible differences in model performance (evaluted by mean squared error). The penalty parameter $\lambda$ was optimized through 5-fold cross-validation (use of 10-fold cross-validation did not meaningfully change the results). For each model, only the most variable 100,000 CpGs according to median absolute deviation (~25% of all available sites shared across platforms) were included in order to decrease the computational burden and ensure that the selected CpGs would have meaningful interindividual variation.

The cross-study learner (CSL) was constructed as an ensemble of study-specific regression models. Scores from each single-study learner (SSL) were combined using the "stacking" approach [@Patil2018], implemented as follows. First, predictions from each SSL to both itself and the other training datasets were combined into a design matrix (with dimensions $N_{total}$ x # SSLs). This formed the input to an additional penalized Cox regression (ridge regression with $\lambda$ optimized through 5-fold CV and coefficients restricted to be non-negative) of all training studies at once. Coefficients from this regression, corresponding to input study-specific SSLs, were normalized to sum to one to produce the CSL weights. For use in new datasets, SSL scores were each standardized to mean zero and unit variance before calculating their weighted sum (using the "stacking" weights) as the final CSL score.

A series of approaches for combining information across cohorts were tested as alternatives to the CSL. The naive "combined" approach consisted of simply aggregating observations from all training sets into a single dataset and training an elastic net regression as described above while adjusting for study as a fixed effect.  The ComBat method trained across all studies as with the “combined” approach, but included an empirical Bayes-based preprocessing step to directly remove mean differences across studies that were not associated with the outcome of interest (incident CVD events) [@Johnson2007].

MRS evaluation in FHS-UM was performed using Cox proportional hazards models, with a series of models adjusting for covariates including demographics, anthropometrics, biochemical values, and cell subtype estimates. Additional sensitivity models incorporated flexible spline bases for age and cell type fractions (*pspline* function) and an interaction between age and sex. Robust standard errors were used to account for family structure as has been suggested for clustered data [@Rogers1993] and used for epigenetic risk models in FHS [@Lu2019]. The proportional hazards assumption was assessed using the *cox.zph* R function, and no violation was detected (p > 0.05). To compare risk scores generated using different models (combined and ComBat-preprocessed) to the CSL, Cox regressions adjusting for the "basic" covariate set were used to evaluate each MRS alone, the CSL MRS plus the combined MRS, and the CSL MRS plus the ComBat-preprocessed MRS in the held-out FHS-UM dataset. Likelihood ratio tests were then used to compare each of the two-MRS models to that CSL-only model, with the resulting p-values indicating whether either of these alternative scores provided additional benefit. MRS evaluation in the REGICOR case-control used logistic regression models, adjusting for the same sets of covariates where possible, though traditional biochemical risk factors were only available in discrete low vs. high categories.

The biology underlying the CSL model was evaluated through a series of enrichment tests using the component CpG loci and annotated genes. Gene ontology-based enrichment analysis of each cohort-specific model was performed using the gometh function from the *missMethyl* package for R [@Phipson2015]. This procedure uses gene annotations for CpGs from the HumanMethylation450 microarray annotation from Illumina (v1.0 B2). Enrichment analysis is then performed for each gene ontology category using Wallenius' noncentral hypergeometric distribution to account for inconsistent representation of CpG sites across genes. The overall merged set of CpGs included in the final CSL model was then tested for enrichment in transcription factor binding sites using HOMER tool [@Heinz2010]. CpG loci (with respect to genome build hg19) were provided as inputs, with 200 base-pair windows and repeat-masked sequences.

## Genomic risk score calculation

Imputed genotype data for WHI were retrieved from dbGaP (accession: phs000746.v2.p3. Variants were filtered for imputation R-squared > 0.3, and annotated with rsIDs, loci, and allelic information using the 1000 Genomes Phase 3 download from dbSNP (download date: April 13, 2018). Weights for the genetic risk score calculation (6,630,151 variants) were based on the genome-wide CVD score developed by Khera et al [@Khera2018]. We note that these scores were developed only for populations of European descent, and thus are not optimized for the mixed-ancestry WHI population. GRS were then calculated as the weighted sum of allelic dosages, normalized by the number of relevant SNPs available. Genotype data processing and GRS calculation were performed using PLINK 2.0.

## Risk score interaction analysis

Interaction analysis was performed using similar Cox regression models to those above, adjusting for the "basic" set of covariates and using robust standard error estimates. To facilitate visual comparisons, main-effect regressions for the MRS were fitted within risk strata defined by the FRS or GRS separately in each dataset. To obtain overall interaction effect estimates, an interaction between MRS and either FRS or GRS was introduced into a combined regression including all datasets, while allowing stratified baseline hazards (strata() argument to the coxph function). We note that main effects in the interaction analysis are biased away from the null since the regression datasets were used for training the MRS. Regressions assessing the GRS excluded non-white participants to match the ancestry used to develop the CVD score [@Khera2018].

For quasi-replication of these associations in the REGICOR dataset, stratified logistic regressions were used to discriminate MI cases from controls using the MRS, while adjusting for estimated cell count fractions as well as two SVA components (as in the main REGICOR models). <!--Age strata comprised three groups: less than 55 years old, 55-65 years old, and greater than 65 years old.--> In the absence of continuous values for blood pressure and lipids, an empirical risk function was generated by first performing a logistic regression on the following cardiovascular risk factors: age, sex, estimated cell count fractions, BMI, diabetes, smoking status, hyperlipidemia (binary), and hypertension (binary), along with two SVA components. Predicted risks based on this model were then used to stratify subjects into four risk groups by evenly splitting the range of predicted risks into four segments (thus resulting in strata based on raw risk, rather than percentiles).

# Results

## Cross-study learner model development

Epigenomic model development was performed in three cohorts, including the Women's Health Initiative (WHI), Framingham Heart Study Offspring Cohort (FHS), and Lothian Birth Cohort 1936 (LBC). The FHS dataset was divided into two functionally separate groups (FHS-JHU and FHS-UM) based on differences in subject selection and geographic location of laboratory methylation analysis (see Methods). Further population details can be found in Table 1.

```{r pop-description}
get_median_iqr <- function(x) {
  if (all(is.na(x))) return (as.character(NA))
  median_fmt <- round(median(x, na.rm=T), 1)
  iqr_borders <- round(quantile(x, c(0.25, 0.75), na.rm=T), 1)
  paste0(median_fmt, " (", iqr_borders[1], "-", iqr_borders[2], ")")
}

study_names_pretty <- c(whi="WHI", fhs_JHU="FHS-JHU", lbc36="LBC", 
                        fhs_UofMinn="FHS-UM")

pop_summary_tbl <- nmd_all %>%
  group_by(batch) %>%
  summarise(`Sample size`=n(),
            Age=get_median_iqr(age),
            `Sex (female)`=paste0(sum(sex == "F"), " (",
                                  round(sum(sex == "F") / n() * 100), "%)"),
            `% European`=paste0(sum(race == "white"), " (",
                                round(sum(race == "white") / n() * 100), "%)"),
            `% African American`=paste0(sum(race == "black"), " (",
                                  round(sum(race == "black") / n() * 100), "%)"),
            `% Hispanic`=paste0(sum(race == "hispanic"), " (",
                                  round(sum(race == "hispanic") / n() * 100), "%)"),
            `Body mass index (kg/m^2)`=get_median_iqr(bmi),
            `LDL cholesterol (mg/dL)`=get_median_iqr(ldl),
            `HDL cholesterol (mg/dL)`=get_median_iqr(hdl),
            `Triglycerides (mg/dL)`=get_median_iqr(tg),
            `Fasting glucose (mg/dL)`=get_median_iqr(glu),
            `Systolic blood pressure (mm Hg)`=get_median_iqr(sbp),
            `Prior only`=sum(pastEvent & !event),
            `Incident only`=sum(!pastEvent & event),
            `Prior and incident`=sum(pastEvent & event),
            Total=sum(pastEvent | event),
            `Follow-up time (years)`=round(max(time) / 365)) %>%
  mutate(batch=factor(batch,
                      levels=names(study_names_pretty),
                      labels=study_names_pretty)) %>%
  replace_na(list(`Fasting glucose (mg/dL)`="Unavailable")) %>%
  arrange(batch) %>% 
  t()
rownames(pop_summary_tbl)[1] <- "Study/Subset"

kable(pop_summary_tbl, booktabs=T,
      caption="Baseline parameters of the populations used for model development") %>%
  footnote(general=c("* Continuous values shown as: median (interquartile range)",
                     "WHI = Women's Health Initiative, FHS-JHU = Framingham Heart Study Offspring Cohort (Johns Hopkins University subset), LBC = Lothian Birth Cohorts 1936, FHS-UM = Framingham Heart Study Offspring Cohort (University of Minnesota subset)"),
           general_title="",
           threeparttable=T) %>%
  pack_rows("Ancestry", 5, 7, bold=F, latex_gap_space="0em") %>%
  pack_rows("# CVD events", 14, 17, bold=F, latex_gap_space="0em")
```

Fig. 1 outlines the computational workflow. Briefly, a cross-study learning (CSL) model was developed by training time-to-event elastic net regressions on three of the datasets, while holding out the FHS-UM subset for evaluation. The FHS-UM subset was chosen to hold out as it more closely represents the larger free-living Framingham population. While there is moderate heterogeneity between the included cohorts (for example, in original cohort study designs, details of CVD definitions, and length of follow-up), the intent of the present investigation was to explore the extraction of shared signal across cohorts with recognized heterogeneity. Next, a model re-trained on all four datasets was subject to external replication in the REGICOR study. CSL model CpGs were characterized as to their potential biological function, and model performance was assessed across strata of alternative cardiovascular risk metrics.

The initial predictor was developed by training individual penalized Cox proportional hazards regression models (single-study learners, or SSLs) in each of the three training cohorts (WHI, FHS-JHU, and LBC). Scores from these models were aggregated through a "stacking" method, in which the outcomes and model predictions from each of the individual datasets are combined, and a regression is used to assign weights to each of the model scores (see Methods). This procedure led to FHS-JHU dropping out of the ensemble model, with weights for this initial predictor as follows: 0.57 (WHI), 0.0 (FHS-JHU), and 0.43 (LBC). This result means that the FHS-JHU score did not transfer to the rest of the datasets (i.e. to WHI and LBC) as well as the scores from the other two components models. 

```{r workflow-diagram, fig.cap="Computational workflow for MRS development and evaluation. The initial MRS was trained in three cohorts with FHS-UM held out to evaluate performance. The final MRS was then trained using all four datasets and examined for biological significance, before testing for prevalent MI discrimination in an independent cohort and assessment of interactions with genetic and traditional risk scores."}
include_graphics("workflow.pdf")
```

## Assessment in held-out FHS subset

```{r fhs-holdout}
pretty_models <- c(unadjusted="Unadjusted",
                   basic="Basic",
                   plus_risk_factors="Plus risk factors",
                   FRS_only="FRS only")
pretty_models_with_ss <- setNames(
  paste0(pretty_models, "\\textsuperscript{", 1:4, "}"),
  names(pretty_models))

model_names_for_tbls <- c(unadjusted="Unadjusted\\textsuperscript{1}",
                          basic="Basic\\textsuperscript{2}",
                          plus_risk_factors="Plus risk factors\\textsuperscript{3}",
                          FRS_only="FRS only\\textsuperscript{4}")

uofm_holdout_res_tbl %>%
  mutate(Model=pretty_models_with_ss[covariate_set],
         HR_per_SD=round(HR_per_SD, 2),
         CI_lower=exp(log(HR_per_SD) + 1.96 * (log(SE_lower) - log(HR_per_SD))),
         CI_upper=exp(log(HR_per_SD) + 1.96 * (log(SE_higher) - log(HR_per_SD))),
         HR_per_SD_fmt=paste0(HR_per_SD, " [", round(CI_lower, 2), "-",
                              round(CI_upper, 2), "]"),
         p=format(p, scientific=T, digits=2)) %>%
  select(Model, HR_per_SD_fmt, p) %>%
  setNames(c("Model", "HR per s.d. MRS\\textsuperscript{a}", "P-value")) %>%
  kable(booktabs=T, caption="MRS performance in held-out FHS subset", escape=F) %>%
  footnote(alphabet="Estimated hazard ratio per standard deviation of the methylation-based risk score",
           number=c("No covariates",
                    "Adjusted for age, sex, and estimated cell type fractions",
                    "Additionally adjusted for BMI, LDL, HDL, SBP, diabetes status, and current smoking",
                    "Adjusted for Framingham Risk Score only (uses all risk factors other than BMI and cell type fractions)"),
           footnote_order=c("alphabet", "number"),
           threeparttable=T)

nmd_with_quintiles <- nmd_all %>%
  filter(batch == "fhs_UofMinn") %>%
  mutate(mrs_quintile=cut(mrs_uofm_holdout, 
                          breaks=quantile(mrs_uofm_holdout, seq(0, 1, by=0.2)),
                          labels=1:5)) %>%
  filter(mrs_quintile %in% c(1, 5)) %>%
  mutate(mrs_quintile=as.character(mrs_quintile))
uofm_holdout_res_tbl_by_quintiles <- coxph(
  Surv(time, event) ~ mrs_quintile + age + sex + CD4T + CD8T + NK + Bcell + Mono + bmi + diabetes + smk_now + ldl + hdl + sbp,
  data=nmd_with_quintiles,
  robust=T) %>%
  tidy()

# compare_df %>%
#   mutate(CI_lower=exp(log(HR_per_SD) + 1.96 * (log(SE_lower) - log(HR_per_SD))),
#          CI_upper=exp(log(HR_per_SD) + 1.96 * (log(SE_higher) - log(HR_per_SD)))) %>%
#   mutate_at(vars(-covariate_set, -Method), round, 2) %>%
#   mutate(res_string=paste0(HR_per_SD, " [", CI_lower, "-", CI_upper, "]"),
#          covariate_set=pretty_models[covariate_set]) %>%
#   select(Method, covariate_set, res_string) %>%
#   spread(key=Method, value=res_string) %>%
#   dplyr::slice(match(c("Unadjusted", "Basic", "Plus risk factors", "FRS only"),
#               covariate_set)) %>%
#   select(covariate_set, combined, combat_nouofm, CSL) %>%
#   setNames(c("Model",
#              "Combined",
#              "ComBat", "CSL")) %>%
#   kable(booktabs=T, escape=F,
#         caption="Comparison of Cox regression coefficient estimates between the CSL and combined model options.") %>%
#   add_header_above(c("", "HR per s.d. MRS\\\\textsuperscript{a}"=3), escape=F) %>%
#   footnote(alphabet="Estimated hazard ratio per standard deviation of the methylation-based risk score",
#            general=c("* Results are presented as: HR per s.d. MRS [95% CI]",
#                      "* Combined model naively combines all datasets while adjusting for study, whereas ComBat adjusts the data themselves to remove variation across studies while preserving class differences.",
#                      "* Model covariates as in Table 2"), 
#            general_title="",
#            footnote_order=c("alphabet", "general"),
#            threeparttable=T)
```

```{r sensitivity-models}
nmd_fhsUM <- filter(nmd_all, batch == "fhs_UofMinn")

# Test proportional hazards assumption in main FHS-UM test
ph_test <- cox.zph(
  coxph(Surv(time, event) ~ mrs_uofm_holdout + age + sex + CD4T + CD8T + NK + Bcell + Mono, 
        data=nmd_fhsUM, robust=T))

# # Test in FHS-UM while excluding all individuals with past events
# no_PE_holdout <- coxph(
#   Surv(time, event) ~ mrs_uofm_holdout, 
#   data=filter(nmd_fhsUM, pastEvent==F),
#   robust=T) %>% 
#   tidy()

# Test in FHS-UM as a binary outcome rather than survival
glm_holdout_mrs <- glm(event ~ mrs_uofm_holdout, 
                   family="binomial", data=nmd_fhsUM)
binary_mrs_preds <- predict(glm_holdout_mrs, newdata=nmd_fhsUM)
mrs_roc <- pROC::roc(nmd_fhsUM$event, binary_mrs_preds, ci=T)
glm_holdout_frs <- glm(event ~ frs, 
                       family="binomial", data=nmd_fhsUM)
binary_frs_preds <- predict(glm_holdout_frs, newdata=nmd_fhsUM)
frs_roc <- pROC::roc(nmd_fhsUM$event, binary_frs_preds, ci=T)
glm_holdout_both <- glm(event ~ frs + mrs_uofm_holdout, 
                       family="binomial", data=nmd_fhsUM)
binary_both_preds <- predict(glm_holdout_both, newdata=nmd_fhsUM)
both_roc <- pROC::roc(nmd_fhsUM$event, binary_both_preds, ci=T)

# Test for residual confounding -- age splines, sex interactions, and cell type fractions
age_spline_test <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout + pspline(age) + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=nmd_fhsUM, robust=T
)
age_sex_interaction_test <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout + age*sex + CD4T + CD8T + NK + Bcell + Mono,
  data=nmd_fhsUM, robust=T
)
cell_type_splines_test <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout + age + sex + pspline(CD4T) + pspline(CD8T) + pspline(NK) + pspline(Bcell) + pspline(Mono),
  data=nmd_fhsUM, robust=T
)
all_resid_confounding_test <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout + pspline(age) + age * sex + pspline(CD4T) + pspline(CD8T) + pspline(NK) + pspline(Bcell) + pspline(Mono),
  data=nmd_fhsUM, robust=T
)

# Sex-stratified evaluations
males_holdout <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout,
  data=filter(nmd_fhsUM, sex == "M"),
  robust=T) %>%
  tidy()
females_holdout <- coxph(
  Surv(time, event) ~ mrs_uofm_holdout,
  data=filter(nmd_fhsUM, sex == "F"),
  robust=T) %>%
  tidy()

# Test models developed excluding stroke
nmd_otheroutcomes <- replace_na(nmd_fhsUM, list(incStroke=F, incCHD=F))
nostroke_mrs_test <- coxph(
  Surv(time, event) ~ scale(mrs_uofm_holdout_nostroke) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=filter(nmd_otheroutcomes),
  robust=T) %>%
  tidy()
nostroke_mrs_test_stroke <- coxph(
  Surv(time, event) ~ scale(mrs_uofm_holdout_nostroke) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=filter(nmd_otheroutcomes, incCHD == F),
  robust=T) %>%
  tidy()
nostroke_mrs_test_chd <- coxph(
  Surv(time, event) ~ scale(mrs_uofm_holdout_nostroke) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=filter(nmd_otheroutcomes, incStroke == F),
  robust=T) %>%
  tidy()

# Test only stroke vs. CHD
mrs_test_stroke <- coxph(
  Surv(time, event) ~ scale(mrs_uofm_holdout) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=filter(nmd_otheroutcomes, incCHD == F),
  robust=T) %>%
  tidy()
mrs_test_chd <- coxph(
  Surv(time, event) ~ scale(mrs_uofm_holdout) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  data=filter(nmd_otheroutcomes, incStroke == F),
  robust=T) %>%
  tidy()

generate_CI <- function(res_row) {
  nonexp_coef <- log(res_row[2])
  nonexp_SE <- log(res_row[2]) - log(res_row[4])
  c(exp(nonexp_coef - 1.96 * nonexp_SE), exp(nonexp_coef + 1.96 * nonexp_SE))
}

# Likelihood ratio tests of combined and ComBat models over CSL only
csl_only_model <- coxph(Surv(time, event) ~ mrs_uofm_holdout + age + sex + CD4T + CD8T + NK + Bcell + Mono,
                        data=nmd_fhsUM, robust=T)
combined_only_model <- coxph(Surv(time, event) ~ mrs_uofm_holdout_combined + age + sex + CD4T + CD8T + NK + Bcell + Mono,
                             data=nmd_fhsUM, robust=T)
combined_plus_csl_model <- coxph(Surv(time, event) ~ mrs_uofm_holdout_combined + mrs_uofm_holdout + age + sex + CD4T + CD8T + NK + Bcell + Mono,
                             data=nmd_fhsUM, robust=T)
combined_csl_lrt <- lmtest::lrtest(csl_only_model, combined_plus_csl_model)
combat_only_model <- coxph(Surv(time, event) ~ mrs_uofm_holdout_combat_nouofm + age + sex + CD4T + CD8T + NK + Bcell + Mono,
                           data=nmd_fhsUM, robust=T)
combat_plus_csl_model <- coxph(Surv(time, event) ~ mrs_uofm_holdout_combat_nouofm + mrs_uofm_holdout + age + sex + CD4T + CD8T + NK + Bcell + Mono,
                               data=nmd_fhsUM, robust=T)
combat_csl_lrt <- lmtest::lrtest(csl_only_model, combat_plus_csl_model)
```

```{r km-plot, warning=F, fig.cap="Kaplan-Meier survival curves in the held-out FHS-UM dataset. Individual curves correspond to tertiles of the initial (3-dataset) MRS. Vertical ticks correspond to censored observations, and colored bands represent 95% confidence intervals for tertile-specific survival curves. X-axis is limited to the time span in which at least 50 uncensored observations remained for each tertile (3275 days)."}
surv_data <- nmd_all %>%
  mutate(test_mrs=mrs_uofm_holdout,
         test_mrs_tertile=cut(test_mrs, quantile(test_mrs, seq(0, 1, length.out=4)),
                              include.lowest=T)) %>%
  filter(batch == "fhs_UofMinn")
uofm_holdout_surv_fit <- survfit(Surv(time, event) ~ test_mrs_tertile,
                                 data=surv_data)
ggsurvplot(uofm_holdout_surv_fit, 
           ylim=c(0.7, NA),
           xlim=c(0, 3275),
           conf.int=T,
           break.time.by=730.5, xscale=365.25, 
           xlab="Time (years)", ylab="Fraction event-free",
           legend=c(0.3, 0.3), legend.title="",
           legend.labs=paste("MRS Tertile", 1:3))
```

Stacking of the three initial predictors resulted in model weights of `r round(uofm_holdout_csl$csl_weights["whi"], 2)`, `r round(uofm_holdout_csl$csl_weights["fhs_JHU"], 2)`, and `r round(uofm_holdout_csl$csl_weights["lbc36"], 2)` for WHI, FHS-JHU, and LBC, respectively (i.e. the FHS-JHU sub-model did not contribute to the initial stacked ensemble model). The resulting ensemble predictor was evaluated using robust Cox proportional hazards models in FHS-UM, showing strong associations with incident CVD in an unadjusted model (HR=1.58, 95% CI: 1.37-1.83), which was attenuated partially through adjustment for standard covariates (age, sex, and estimated cell type fractions; HR=1.28, 95% CI: 1.10-1.50) as well as CVD risk factors (HR=1.29, 95% CI: 1.09-1.51). Results for the unadjusted model and three risk factor-adjusted models are shown in Table 2, and associated Kaplan-Meier curves across epigenetic risk tertiles are shown in Fig. 2.

Additional sensitivity analyses were performed to assess the robustness of these results to variations in the model-building or evaluation approach. Hazard ratios in the held-out FHS-UM were no higher using penalized logistic regression in training (unadjusted HR=1.52, 95% CI: 1.32-1.76), excluding individuals with past events in training (unadjusted HR=1.55, 95% CI: 1.33-1.81), or adjusting for race in WHI (unadjusted HR=1.20, 95% CI: 1.03-1.39). Neither were these results affected by training using the full set of 390,597 CpGs. Similarly, variations in the evaluation regressions did not produce meaningfully different results, either when excluding all individuals who experienced prior CVD events (Supp. Table S1), analyzing incident CVD as a binary outcome using logistic regression (unadjusted odds ratio per standard deviation=2.15, 95% CI: 1.91-2.42), or stratifying by sex. Adjustment for age and cell type fractions as flexible spline functions as well as an age-sex interaction to assess possible residual confounding did not decrease estimated HRs from the Basic model (saturated model HR=1.31, 95% CI: 1.12-1.52). Use of the MRS for binary incident CVD prediction resulted in a c-statistic of 0.642 (95% CI: 0.599-0.685), compared to 0.691 (95% CI: 0.653-0.729) for the Framingham Risk Score alone and 0.695 (95% CI: 0.655-0.734) using the two scores together.

Results from comparison of CSL performance to models trained on combined datasets (either naive combination or including preprocessing using ComBat) are shown in Supp. Fig. S1. The ComBat-preprocessed model had modestly higher hazard ratios in FHS-UM, while relative differences with the combined model depended on the covariates included. However, likelihood ratio tests using the Basic model covariates (age, sex, and cell type fraction-adjusted) did not reveal a strong added benefit of either the combined (p = 0.58) or ComBat (p = 0.08) risk scores over that using only the CSL.

## Final CSL model characterization

```{r overlap}
venn_order <- c("whi", "fhs_UofMinn", "fhs_JHU", "lbc36")
silent <- VennDiagram::venn.diagram(
  lapply(ssl_nonzero_coefs, names)[venn_order],  # Ordering
  "model_cpgs_venn.png", 
  category.names=study_names_pretty[venn_order],
  cat.cex=2, cex=2,
  cat.just=list(c(0.2,0), c(0.6,0), c(0.4,0), c(0.3,0)),
  fill=c('red', 'green', 'blue', 'purple'),
  imagetype="png")
# venn diagram formatting reference: https://www.r-graph-gallery.com/14-venn-diagramm/

venn_img <- grid::rasterGrob(png::readPNG("model_cpgs_venn.png"), 
                             interpolate=T)


# ssl_nonzero_names <- lapply(ssl_nonzero_coefs, names)
# ssl_nonzero_annos <- lapply(ssl_nonzero_names, function(nms) {
#   anno_450k %>%
#     filter(Name %in% nms) %>%
#     select(Name, chr, pos)
# })
# ssl_nonzero_ranges <- lapply(ssl_nonzero_annos, function(anno) {
#   GRanges(anno$chr, IRanges(anno$pos - 1000, anno$pos + 1000))
# })
```

```{r gsa, cache=1, message=F, warning=F, eval=F}
gsa_tbls <- lapply(ssl_nonzero_coefs, function(set) {
  missMethyl::gometh(names(set), all.cpg=common_cpgs,
                     collection="GO", array.type="450K",
                     anno=anno_450k) %>%
    arrange(P.DE)
})
```

```{r annotation-enrichment, cache=1, message=F, warning=F}
summarise_annotation <- function(group, df) {
  # Takes in a set of annotations and returns the proportions of a given field (e.g. gene name)
  df %>%
    select(one_of("Name", group)) %>%
    separate_rows(group, sep=";") %>%
    distinct() %>%
    filter_at(group, all_vars(.!="")) %>%
    group_by_at(group) %>%
    summarise(n=n()) %>%
    mutate(frac=n/sum(n))
}

calc_enrichment_p <- function(mod_cpgs, bg_cpgs, type, group, lower_tail=F) {
  rti_summary_mod <- summarise_annotation(type, filter(anno_450k, 
                                                       Name %in% mod_cpgs))
  rti_summary_all <- summarise_annotation(type, filter(anno_450k, 
                                                       Name %in% bg_cpgs))
  n_hits_in_group <- sum(rti_summary_mod$n[rti_summary_mod[[type]] %in% group])
  n_avail_in_group <- sum(rti_summary_all$n[rti_summary_all[[type]] %in% group])
  2 * phyper(n_hits_in_group, n_avail_in_group, 
             length(wgcna_input_cpgs) - n_avail_in_group, length(mod_cpgs), 
             lower.tail=lower_tail)
}

annot_plot <- function(cpg_set, bg_cpg_set, group_var) {
  # Given a specific field, plot proportion of CpGs in a specific subset 
  # next to those of the full set
  my_groups <- summarise_annotation(
    group_var, filter(anno_450k, Name %in% cpg_set))
  all_groups <- summarise_annotation(
    group_var, filter(anno_450k, Name %in% bg_cpg_set))
  merge_groups <- inner_join(my_groups, all_groups, by=group_var, 
                             suffix=c(".subset", ".all")) %>%
    gather(key=group, value=frac, contains("frac")) %>%
    mutate(group=factor(group, levels=c("frac.all", "frac.subset"), 
                        labels=c("All CpGs", "MRS CpGs")))
  ggplot(merge_groups, aes_string(x=group_var, y="frac", fill="group")) +
    geom_bar(stat="identity", position="dodge") +
    theme(axis.title.x=element_blank(), 
          axis.text.x=element_text(angle=35, hjust=1))
}

annot_groups <- function(cpg_set, bg_cpg_set1, bg_cpg_set2, group_var) {
  # Given a specific field, plot proportion of CpGs in a specific subset 
  # next to those of the full set
  my_set <- summarise_annotation(
    group_var, filter(anno_450k, Name %in% cpg_set))
  var_set <- summarise_annotation(
    group_var, filter(anno_450k, Name %in% bg_cpg_set1))
  all_cpgs <- summarise_annotation(
    group_var, filter(anno_450k, Name %in% bg_cpg_set2))
  merge_groups <- bind_rows(list(myset=my_set,
                                 bg1=var_set,
                                 bg2=all_cpgs),
                            .id="set") %>%
    mutate(set=factor(set,
                      levels=c("myset", "bg1", "bg2"),
                      labels=c("MRS CpGs", "100k variable CpGs", "All CpGs")))
  # merge_groups <- inner_join(my_set, all_cpgs, by=group_var, 
  #                            suffix=c(".subset", ".all")) %>%
  #   gather(key=set, value=frac, contains("frac")) %>%
  #   mutate(set=factor(set, levels=c("frac.all", "frac.subset"), 
  #                     labels=c("All CpGs", "Module CpGs")))
  merge_groups$group <- merge_groups[[group_var]]
  merge_groups
}

all_nonzero_cpgs <- Reduce(union, lapply(ssl_nonzero_coefs, names))

csl_gene_group_enrichment <- annot_groups(
  all_nonzero_cpgs, variable_100k_cpgs, common_cpgs, "UCSC_RefGene_Group")
csl_cpg_island_enrichment <- annot_groups(
  all_nonzero_cpgs, variable_100k_cpgs, common_cpgs, "Relation_to_Island")
```

```{r characterization, fig.asp=1.3, fig.cap="Characterization of the final CSL model. a) Overlap of CpG sites in the four individual predictors constituting the final model. b) Study-specific weights for constructing the ensemble model (derived from the \"stacking\" regression). c) Results from Gene Ontology-based enrichment analysis using genes annotated to SSL component CpGs. All GO terms with false discovery rate < 0.001 in any cohort are shown, and colored according to -log(p-value) for enrichment in each SSL. Values were cut at -log(p) = 20 for visualization purposes. d) Proportion of CpGs in the full set of CSL CpGs (union of CpG sets in each component SSL) compared to the 100,000 most variable CpGs (as used in SSL model development) and the full set of available CpGs. Groupings according to both gene-based and CpG island-based CpG annotations are shown."}
# gsa_summary <- bind_rows(lapply(gsa_tbls, function(x) x[1, ]), .id="Study") %>%
#   mutate(diff_meth_genes=paste0(DE, "/", N),
#          P.DE=format(P.DE, scientific=T, digits=2),
#          FDR=format(FDR, scientific=T, digits=2)) %>%
#   select(Study, Term, diff_meth_genes, P.DE, FDR) %>%
#   mutate(Study=study_names_pretty[Study]) %>%
#   setNames(c("Study", "Biological Process", "# Relevant Genes", "P-value", "FDR")) %>%
#   as_tibble()

fdr_processes <- unlist(
  lapply(gsa_tbls, function(tbl) filter(tbl, FDR < 0.001, 
                                        Ont == "BP")$Term))

term_replace <- c(
  "Cell-cell adhesion via plasma-membrane adhesion molecules"="Cell-cell adhesion via plasma-\nmembraneadhesion molecules",
  "Homophilic cell adhesion via plasma membrane adhesion molecules"="Homophilic cell adhesion via plasma-\nmembrane adhesion molecules\n",
  "Multicellular organism development"="Multicellular organism\ndevelopment",
  "Nervous system development"="Nervous system\ndevelopment"
)

gsa_summary_tbl <- bind_rows(gsa_tbls, .id="Study") %>%
  filter(Term %in% fdr_processes) %>% 
  mutate(negLogP=-log(P.DE),
         Study=factor(Study,
                      levels=names(study_names_pretty),
                      labels=study_names_pretty)) %>%
  complete(Study, Term, fill=list(FDR=NA, P.DE=NA)) %>%
  arrange(P.DE) %>%
  mutate(Term=str_to_sentence(Term),
         Term=ifelse(Term %in% names(term_replace), term_replace[Term], Term),
         Term=fct_rev(factor(Term, levels=unique(Term))))

gsa_summary_plt <- gsa_summary_tbl %>%
  mutate(negLogP=pmin(20, negLogP)) %>%
  ggplot(aes(x=Study, y=Term, fill=negLogP)) +
  geom_tile(color="black", width=0.75, height=0.75, size=0.8) +
  scale_fill_gradient(low="gray", high="red", name="-log(p)",
                      breaks=c(5, 10, 15, 20), labels=c(5, 10, 15, ">20")) +
  scale_x_discrete(position="top") +
  scale_y_discrete(position="right") +
  theme(axis.line=element_blank(), axis.ticks=element_blank(),
        axis.title=element_blank(), 
        legend.position=c(1.6, 0.5),
        # legend.key.width=unit(0.06, "npc"),
        axis.text.x=element_text(angle=60, hjust=0, size=10),
        axis.text.y=element_text(size=7))

# gsa_summary_img <- gridExtra::tableGrob(
#   gsa_summary, rows=NULL, theme=gridExtra::ttheme_default(base_size=6))

weights_plt <- data.frame(
  study=factor(study_names_pretty, levels=study_names_pretty),
  weight=final_csl$csl_weights[names(study_names_pretty)]) %>%
  ggplot(aes(x=fct_rev(study), y=weight), width=0.1) +
  geom_bar(stat="identity") +
  labs(y="Ensemble weight") +
  coord_flip() +
  theme(axis.title.y=element_blank(), 
        axis.ticks.y=element_blank(),
        axis.line.x=element_blank())

annot_plt <- bind_rows(list(gene=csl_gene_group_enrichment, 
                            cpg_island=csl_cpg_island_enrichment), 
                       .id="type") %>%
  select(-UCSC_RefGene_Group, -Relation_to_Island) %>%
  na.omit() %>% 
  mutate(type=factor(type, levels=c("gene", "cpg_island"),
                     labels=c("Gene-based", "CpG Island-based")),
         group=gsub("_", ". ", group)) %>%
  ggplot(aes(x=group, y=frac, fill=set)) +
  geom_bar(stat="identity", position="dodge") +
  labs(y="CpG fraction with annotation") +
  scale_fill_brewer(palette="Dark2") +
  facet_wrap(~type, scales="free_x") +
  theme(axis.title.x=element_blank(), 
        axis.text.x=element_text(angle=35, hjust=1, size=9.5),
        axis.title.y=element_text(size=9.5),
        axis.line.y=element_blank(),
        legend.title=element_blank(), 
        strip.background=element_blank())

top_left <- plot_grid(venn_img, weights_plt, nrow=2, labels=c("a", "b"))

top_row <- plot_grid(top_left, gsa_summary_plt, ncol=2, 
                     rel_widths=c(2, 3), labels=c("", "c"))

plot_grid(top_row, annot_plt, nrow=2, rel_heights=c(3, 2), labels=c("", "d"))
```

```{r tf-enrichment, message=F}
homer_res <- read_tsv("../../output/homer_results/csl/knownResults.txt")
```

The stacking regression in the final CSL model defining the methylation-based risk score (MRS) gave the most weight to WHI (0.48) and LBC (0.38), while retaining nonzero weights for FHS-JHU (0.06) and FHS-UM (0.08). This result indicates that the WHI and LBC-trained models were better able to transfer across the combined-cohort set of outcomes compared to the other models. There was little overlap of specific CpG sites across cohort-specific models, with a maximum of 13 CpGs shared between two models (WHI and FHS-UM) and no CpGs shared between three or more models (Fig. 3a). This could result from heterogeneity in the complex relationships between DNA methylation and CVD across populations. However, it may also reflect the tendency of the elastic net regression to select only a single feature from a group of correlated features, where the specific CpGs selected in different datasets varied due to the presence of biological and technical noise. However, even if the SSLs were capturing different biological mechanisms, the CSL model is designed to capture such heterogeneous signal from across cohorts. Despite the lack of site-specific overlap, there was broad agreement for three of the four component SSL models at the level of enriched biological processes, with all except FHS-JHU enriched most strongly for proximity to genes involved in homophilic cell adhesion (Fig. 3b). MRS component CpGs tended to be found in similar genomic loci to the overall set of variable CpGs, and were enriched in gene bodies and depleted in CpG islands compared to the full microarray CpG set. However, MRS CpGs did show a modest enrichment in and around CpG islands compared to the set of variable CpGs (Fig. 3d). To seek more clarity as to potential biological mechanisms represented by the MRS, the HOMER tool was used to calculate enrichment of transcription factor (TF) binding motifs in the MRS component CpG sites. Using the union of all individual SSL CpG sites as input, no strong enrichments were found (all q-values >0.5).

To better understand the stability of the risk score over time, intraclass correlation coefficients (ICCs) were calculated for two sets of grouped samples: 26 technical replicates from FHS and approximately 1000 longitudinal samples (across 3 visits, or about 6 years total) from LBC (Supp. Table S2). The technical replicates showed an ICC of 0.85, while the longitudinal samples showed an ICC of 0.68. As would be expected, the ICC for samples closer in time (Waves 1 & 2; ICC = 0.69) were higher than that for samples more distant in time (Waves 1 & 3; ICC = 0.61). Based on the observation of imperfect stability of the MRS over time as well as the partial attenuation in held-out hazard ratios after adjustment for age, its component CpGs (the `r length(all_nonzero_cpgs)`-element union of all CpGs in any of the four individual SSL models) were examined for overlap with established epigenetic age metrics. While no enrichment was seen for the original cross-tissue DNAm age from Horvath [@Horvath2013], strong enrichment was seen for the morbidity-directed PhenoAge [@Levine2018] (9 of 513 CpGs; p=2.3e-5) and especially the blood-specific aging marker from Hannum et al. [@Hannum2013] (13 of 71 CpGs; p=5.9e-21). We note that these overlaps do not constitute a major fraction of either CpG set, but are nonetheless highly statistically significant. The PhenoAge metric is based on some known cardiovascular risk factors (e.g. C-reactive protein) and is known to associate with CVD, but is not trained in any of the cohorts included here.

## Discrimination in myocardial infarction case-control study

```{r regicor, warning=F}
# combat_model_results <- list(
#   unadjusted=as.matrix(data.frame(Estimate=0.5352, `Std. Error`=0.1358, `Pr(>|z|)`=8.058e-05,
#                         row.names="mrs", check.names=F)),
#   basic=as.matrix(data.frame(Estimate=0.7453, `Std. Error`=0.1663, `Pr(>|z|)`=7.427e-06,
#                    row.names="mrs", check.names=F)),
#   plus_risk_factors=as.matrix(data.frame(Estimate=0.5866, `Std. Error`=0.2018, `Pr(>|z|)`=3.651e-03,
#                                row.names="mrs", check.names=F))
# )
# 
# load("../../output/regicor_results_20190221/main_MRS_results.RData")
# delong_test <- readRDS("../../output/regicor_results_20190221/delong_test.rds")
# load("../../output/regicor_results_20190221/stratified_results.RData")

load("../../output/regicor_results_august2019/main_mrs_results.RData")

csl_res_regicor <- lapply(csl_model_results, 
                          function(tbl) as.data.frame(tbl["mrs", , drop=F])) %>%
  bind_rows(.id="Model")
combined_res_regicor <- lapply(combined_model_results, 
                               function(tbl) as.data.frame(tbl["mrs", , drop=F])) %>%
  bind_rows(.id="Model")
combat_res_regicor <- lapply(combat_model_results, 
                             function(tbl) as.data.frame(tbl["mrs", , drop=F])) %>%
  bind_rows(.id="Model")

regicor_res_tbl <- bind_rows(Combined=combined_res_regicor,
                             ComBat=combat_res_regicor,
                             CSL=csl_res_regicor, 
                             .id="model_type") %>%
  dplyr::rename(p=`Pr(>|z|)`,
                SE=`Std. Error`) %>%
  mutate(Model=factor(Model, levels=names(pretty_models), labels=pretty_models),
         OR_per_SD=exp(Estimate),
         CI_lower=exp(Estimate - 1.96 * SE),
         CI_upper=exp(Estimate + 1.96 * SE)) %>%
  mutate_at(vars(OR_per_SD, CI_lower, CI_upper), round, 2) %>%
  mutate(res_string=paste0(OR_per_SD, " [", CI_lower, "-", CI_upper, "]")) %>%
  # mutate(res_string=paste0(round(OR_per_SD, 2), " (", 
  #                          format(p, digits=3, scientific=T), ")")) %>%)
  select(model_type, Model, res_string) %>%
  spread(key=model_type, value=res_string)

kable(regicor_res_tbl, booktabs=T,
      caption="Results from replication in REGICOR MI case-control") %>%
  footnote(general=c("* Results are presented as: OR per s.d. MRS [95% CI]",
                     "* Model covariates as in Table 2",
                     "* All models above are adjusted for two surrogate variable analysis (SVA) components."), 
           general_title="",
           threeparttable=T)
```

As one form of replication, the MRS was investigated for its discriminative performance in a nested case-control for prior myocardial infarction in the REGICOR cohort (Table 3; cohort description in Supp. Table S3), which was matched for sex and age and thus free of potential confounding by these variables. We note that this dataset contained prevalent (rather than incident) events, and thus provides replication in a similar but not identical biological context. These methylation data were collected on the next generation of Illumina methylation microarray (MethylationEPIC), which does not perfectly overlap with the HumanMethylation450 platform, but contained approximately 93% of the CpGs input to the MRS model training procedure. The MRS was able to discriminate cases and controls in both unadjusted (odds ratio = 1.79, p = 6.33e-6) and, to a lesser degree, risk factor-adjusted models (odds ratio = 1.61, p = 0.019). Odds ratios were qualitatively similar across modeling strategies (Combined, ComBat, and CSL) for all of the adjustment models (Supp. Fig. S1b).

## Interactions with alternate risk metrics

```{r aggregate-interactions, warning=F, fig.cap="Interactions of MRS with other biomarkers of CVD risk. a) Hazard ratios for the MRS within subsets of 10-year generalized CVD risk according to the Framingham Risk Score. b) Hazard ratios for the MRS within quartiles of a genetic cardiovascular risk score (in white WHI participants only). Hazard ratios are estimated using the final MRS, which was trained using each of these datasets. Cox regressions included stratum-specific baseline hazards and were adjusted for age, sex, and estimated cell subtype fractions. Error bars represent standard errors for the hazard ratio estimates. Annotated p-values describe the test of interaction between the MRS and the alternative risk metric."}
mrs_frs_interaction_model <- coxph(
  Surv(time, event) ~ scale(mrs_final) * frs + strata(batch) + age + sex + CD4T + CD8T + NK + Bcell + Mono, 
  robust=T, data=nmd_all) %>%
  tidy()
mrs_frs_interaction_effect_per10pct <- round(
  (1 - exp(mrs_frs_interaction_model$estimate[10] / 10)) * 100)
mrs_frs_interaction_pval <- signif(
  mrs_frs_interaction_model$p.value[10], 3)

all_mrs_sd <- sd(nmd_all$mrs_final)
# Note: study-specific estimates (in int_res_list) are implicitly normalized
# by the study-specific MRS std. dev. Here, the overall (combined-study)
# std. dev. is explicitly calculated to normalize estimates.
agg_frs_tbl <- map_dfr(levels(nmd_all$frs_group), function(g) {
  reg_data <- nmd_all %>%
    filter(frs_group == g) %>%
    group_by(batch) %>%
    filter(sum(event) > 25) %>%
    ungroup()
  coxph(Surv(time,event) ~ mrs_final + strata(batch) + age + sex + CD4T + CD8T + Bcell + NK + Mono,
        data=reg_data) %>% tidy() %>% filter(term == "mrs_final")
}, .id="frs_group") %>%
  mutate(frs_group=factor(frs_group, labels=levels(nmd_all$frs_group)[as.integer(frs_group)]),
         estimate=estimate * all_mrs_sd,
         std.error=std.error * all_mrs_sd,         
         HR_per_SD=exp(estimate),
         SE_lower=exp(estimate - std.error),
         SE_higher=exp(estimate + std.error))
agg_frs_plt <- ggplot(agg_frs_tbl, aes(x=frs_group, y=HR_per_SD)) +
  geom_point() +
  geom_errorbar(aes(ymin=SE_lower, ymax=SE_higher), width=0) + 
  annotate("text", label="MRS-FRS interaction:\np = 8.27e-5", x=3, y=2.5, size=3.5) +
  labs(x="FRS 10-year risk") #+
  # theme(axis.text.x=element_text(angle=30, size=10, hjust=0.75))

whi_white_mrs_sd <- sd(filter(nmd_all, study == "whi", race == "white")$mrs_final)
agg_grs_tbl <- map_dfr(levels(nmd_all$grs_group_white), function(g) {
  reg_data <- nmd_all %>%
    filter(race == "white") %>%
    filter(grs_group_white == g) %>%
    filter(batch == "whi") %>%
    group_by(batch) %>%
    filter(sum(event) > 25) %>%
    ungroup()
  # coxph(Surv(time,event) ~ mrs_final + strata(batch) + age + sex + CD4T + CD8T + Bcell + NK + Mono,
  #       data=reg_data) %>% tidy() %>% filter(term == "mrs_final")
  coxph(Surv(time,event) ~ mrs_final + age + CD4T + CD8T + Bcell + NK + Mono,
        data=reg_data) %>% 
    tidy() %>% 
    filter(term == "mrs_final")  
}, .id="grs_group") %>%
  mutate(grs_group=factor(grs_group, labels=levels(nmd_all$grs_group_white)[as.integer(grs_group)]),
         estimate=estimate * whi_white_mrs_sd,
         std.error=std.error * whi_white_mrs_sd,
         HR_per_SD=exp(estimate),
         SE_lower=exp(estimate - std.error),
         SE_higher=exp(estimate + std.error))
agg_grs_plt <- ggplot(agg_grs_tbl, aes(x=grs_group, y=HR_per_SD)) +
  geom_point() +
  geom_errorbar(aes(ymin=SE_lower, ymax=SE_higher), width=0) + 
  annotate("text", label="MRS-GRS interaction:\np = 0.719", x=3, y=2.5, size=3.5) +
  labs(x="Genetic risk score quartile") #+
  # theme(axis.text.x=element_text(angle=30, size=10, hjust=0.75))

# study_legend <- get_legend(frs_quantile_plt)
plot_grid(
  agg_frs_plt + 
    scale_x_discrete(labels=c("<0.1", "0.1-0.2", "0.2-0.4", ">0.4")) +
    labs(y="Hazard ratio (per s.d. MRS)") +
    coord_cartesian(ylim=c(1.65, 2.65)) +
    theme(axis.text.x=element_text(size=9), 
          axis.title=element_text(size=10),
          legend.position="none"),
  agg_grs_plt + theme(legend.position="none") + 
    coord_cartesian(ylim=c(1.65, 2.65)) +
    labs(y="") +
    theme(axis.text.x=element_text(size=9, hjust=0.75), 
          axis.title=element_text(size=10)),
  nrow=1, align="h", axis="bt", 
  labels=c("a", "b"))


g <- coxph(Surv(time,event) ~ mrs_final + strata(batch), data=nmd_all)
mrs_frs_scatterplt <- nmd_all %>%
  filter(mrs_final > quantile(mrs_final, 0.75)) %>%
  ggplot(aes(x=frs, y=mrs_final)) +
  geom_point(aes(alpha=event)) +
  scale_alpha_manual(values=c(0.2, 0.6)) +
  coord_cartesian(ylim=c(0, 3))

mrs_frs_interaction_model <- coxph(
  Surv(time, event) ~ scale(mrs_final) * frs + strata(batch) + age + sex + CD4T + CD8T + NK + Bcell + Mono, 
  robust=T, data=nmd_all) %>%
  tidy()
mrs_frs_interaction_effect_per10pct <- round(
  (1 - exp(mrs_frs_interaction_model$estimate[10] / 10)) * 100)
mrs_frs_interaction_pval <- signif(
  mrs_frs_interaction_model$p.value[10], 3)

mrs_grs_interaction_model <- coxph(
  # Surv(time, event) ~ mrs_final * scale(grs) + strata(batch) + age + sex + CD4T + CD8T + NK + Bcell + Mono,
  Surv(time, event) ~ scale(mrs_final) * scale(grs) + age + CD4T + CD8T + NK + Bcell + Mono,
  robust=T, data=filter(nmd_all, study == "whi", race == "white")) %>%
  tidy()
mrs_grs_interaction_effect_per_sd <- round(
  (1 - exp(mrs_grs_interaction_model$estimate[9])) * 100)
mrs_grs_interaction_pval <- signif(
  mrs_grs_interaction_model$p.value[9], 3)

# fhs_um_model <- coxph(as.formula(paste0("Surv(time, event) ~ mrs_final * grs +",
#                                         paste(setdiff(basic_covars, "sex")))),
#                       data=nmd_all, subset=nmd_all$batch == "fhs_UofMinn")
# fhs_jhu_model <- coxph(as.formula(paste0("Surv(time, event) ~ mrs_final * grs +",
#                                         paste(setdiff(basic_covars, "sex")))),
#                       data=nmd_all, subset=nmd_all$batch == "fhs_JHU")
# whi_model <- coxph(as.formula(paste0("Surv(time, event) ~ mrs_final * grs +",
#                                         paste(setdiff(basic_covars, "sex")))),
#                       data=nmd_all, subset=nmd_all$batch == "whi")
# a <- lapply(list(fhs_um_model, fhs_jhu_model, whi_model, lbc_model),
#        function(model) {
#          broom::tidy(model) %>%
#            filter(term == "mrs_final:grs")
#        }) %>%
#   bind_rows()
# meta::metagen(a$estimate, a$std.error)
```

```{r individual-interactions}
basic_covars <- c("age", "sex", "CD4T", "CD8T", "Bcell", 
                  "NK", "Mono")

test_mrs <- function(meta, mrs, subset=NULL, return_fit=F,
                     covars=c(), robust=F) {
  meta$mrs <- scale(mrs)
  meta$cvd_surv <- Surv(time=meta$time, event=meta$event)  # Outcome
  meta <- mutate_at(meta, c("chol", "ldl", "hdl", "tg", "glu", "hscrp", "sbp"), log10)
  covar_string <- "cvd_surv ~ mrs"
  if (length(covars) > 0) {
    covar_string <- paste(covar_string, "+", paste(covars, collapse=" + "))
  } 
  form <- as.formula(paste(c("cvd_surv ~ mrs", covars), collapse=" + "))
  mrs_test <- coxph(form, data=meta, subset=subset, robust=robust)  # Test the MRS using a Cox model
  mrs_res <- summary(mrs_test)$coef["mrs", c("exp(coef)","z")]
  mrs_res_tbl <- tibble(HR_per_SD=mrs_res["exp(coef)"], 
                        p=2 * pnorm(-abs(mrs_res["z"])))
  if (return_fit) {
    list(fit=mrs_test, tbl=mrs_res_tbl) 
  } else {
    mrs_res_tbl
  }
}

test_mrs_with_SE <- function(nmd, mrs, subset, covars, robust=F) {
  test_res_list <- test_mrs(nmd, mrs, subset=subset, return_fit=T,
                            covars=covars, robust=robust)
  res_tbl <- test_res_list$tbl
  beta <- log(res_tbl$HR_per_SD)
  se <- sqrt(diag(test_res_list$fit$var)[1])
  res_tbl$SE_lower <- exp(beta - se)
  res_tbl$SE_higher <- exp(beta + se)
  res_tbl$n <- test_res_list$fit$n
  res_tbl
}

strat_test_by_study <- function(df, test_mrs, quantile_cat) {
  if (is.factor(df[[quantile_cat]])) {
    unique_levels <- na.omit(levels(df[[quantile_cat]]))
  } else {
    unique_levels <- na.omit(unique(df[[quantile_cat]]))
  }
  df$mrs <- test_mrs
  full_res <- map_dfr(unique_levels, function(q) {
    res_per_study <- df %>%
      nest(-batch) %>%
      mutate(res=map(data, function(d) {
        na_df <- data.frame(HR_per_SD=NA, p=NA, SE_lower=NA, 
                            SE_higher=NA, n=NA)
        subset <- d[[quantile_cat]] == q
        # if (sum(d$event[subset], na.rm=T) < 25) return(na_df)
        tryCatch(test_mrs_with_SE(d, d$mrs, subset, basic_covars),
                 error=function(e) test_mrs_with_SE(
                   d, d$mrs, subset, basic_covars[basic_covars != "sex"]))
      }))
    cbind(study=res_per_study$batch,
          bind_rows(res_per_study$res))
  }, .id="Group")
  full_res$Group <- unique_levels[as.numeric(full_res$Group)]
  full_res
}

int_res_frs <- strat_test_by_study(nmd_all, nmd_all$mrs_final, "frs_group") %>%
  mutate(Group=factor(Group, levels=unique(Group)))
```

```{r regicor-interactions}
# regicor_age_strat <- tibble(
#   `Age group`=c("0-55", "55-65", "66+"),
#   OR_per_SD=c(6.79, 3.43, 1.29),
#   SE_lower=c(3.91, 2.57, 1.03),
#   N=c(64, 161, 166)
# ) %>%
#   mutate(SE=log(OR_per_SD) - log(SE_lower),
#          CI_lower=round(exp(log(OR_per_SD) - 1.96 * SE), 2),
#          CI_upper=round(exp(log(OR_per_SD) + 1.96 * SE), 2),
#          res_string=paste0(OR_per_SD, " [", CI_lower, "-", CI_upper, "]")) %>%
#   select(`Age group`, res_string, N)
# 
regicor_rf_strat <- tibble(
  `Risk factor group`=c("G1", "G2", "G3", "G4"),
  OR_per_SD=c(4.97, 1.12, 2.92, 1.36),
  SE_lower=c(2.88, 0.831, 1.75, 0.639),
  N=c(119, 90, 60, 55)
) %>%
  mutate(SE=log(OR_per_SD) - log(SE_lower),
         CI_lower=round(exp(log(OR_per_SD) - 1.96 * SE), 2),
         CI_upper=round(exp(log(OR_per_SD) + 1.96 * SE), 2),
         res_string=paste0(OR_per_SD, " [", CI_lower, "-", CI_upper, "]")) %>%
  select(`Risk factor group`, res_string, N)
```

```{r testing-123, eval=F}
my_test_mrs <- function(meta, mrs, subset=NULL, return_fit=F,
                     covars=c(), robust=F) {
  meta$mrs <- scale(mrs)
  meta$cvd_surv <- Surv(time=meta$time, event=meta$event)  # Outcome
  form <- as.formula(paste(c("cvd_surv ~ mrs + strata(batch)", covars), collapse=" + "))
  mrs_test <- coxph(form, data=meta, subset=subset, robust=robust)  # Test the MRS using a Cox model
  mrs_res <- summary(mrs_test)$coef["mrs", c("exp(coef)","z")]
  mrs_res_tbl <- tibble(HR_per_SD=mrs_res["exp(coef)"], 
                        p=2 * pnorm(-abs(mrs_res["z"])))
  if (return_fit) {
    list(fit=mrs_test, tbl=mrs_res_tbl) 
  } else {
    mrs_res_tbl
  }
}

my_test_mrs_with_SE <- function(nmd, mrs, subset, covars, robust=F) {
  test_res_list <- my_test_mrs(nmd, mrs, subset=subset, return_fit=T,
                            covars=covars, robust=robust)
  res_tbl <- test_res_list$tbl
  beta <- log(res_tbl$HR_per_SD)
  se <- sqrt(diag(test_res_list$fit$var)[1])
  res_tbl$SE_lower <- exp(beta - se)
  res_tbl$SE_higher <- exp(beta + se)
  res_tbl$n <- test_res_list$fit$n
  res_tbl
}

basic_covars <- c("age", "sex", "CD4T", "CD8T", "Bcell", 
                  "NK", "Mono")

my_strat_test_by_study <- function(df, test_mrs, quantile_cat) {
  df$mrs <- test_mrs
  unique_levels <- levels(df[[quantile_cat]])
  full_res <- map_dfr(setNames(unique_levels, unique_levels), function(q) {
    subset <- df[[quantile_cat]] == q
    my_test_mrs_with_SE(df, df$mrs, subset, basic_covars)
  }, .id="Group")
  full_res
}

my_strat_test_by_study(nmd_all, nmd_all$mrs_final, "frs_group") %>%
  mutate(Group=factor(Group, levels=levels(nmd_all$frs_group))) %>%
  ggplot(aes(x=Group, y=HR_per_SD)) +
  geom_point(position=position_dodge(width=0.4)) +
  geom_errorbar(aes(ymin=SE_lower, ymax=SE_higher), width=0,
                position=position_dodge(width=0.4)) +
  labs(x="FRS Group", title="Showing MRS performance vs. FRS Group when combined")
```

```{r interaction-breakdown, fig.cap="Identification of high-risk individuals in FHS-UM with low traditional risk. a) Scatter plot of the initial MRS (FHS-UM holdout model; arbitrary units) vs. FRS (predicted 10-year risk of a CVD event) in individuals from FHS-UM. b) Fractions of 10-year CVD events in FHS-UM low-risk individuals (FRS < 0.1) below or above the 80th percentile of the 3-cohort MRS. c) As in (b), but in all individuals with FRS < 0.2."}
# ggplot(nmd_all, aes(x=frs, y=mrs_final, color=event)) +
#   geom_point(alpha=0.25) +
#   scale_color_discrete(name="Incident event") +
#   labs(x="FRS 10-year risk", y="Methylation risk score")

rs_compare_data <- nmd_all %>%
  mutate(event_10yr=event & (time <= 10 * 360))

mrs_vs_frs_plt <- rs_compare_data %>%
  filter(batch == "fhs_UofMinn") %>%
  mutate(event_10yr=ifelse(event_10yr, "Yes", "No")) %>%
  ggplot(aes(x=frs, y=mrs_uofm_holdout, color=event_10yr)) +
  geom_point(alpha=0.35) +
  geom_smooth(color="gray", method="lm", se=F) +
  scale_color_brewer(name="Incident event", palette="Dark2") +
  labs(x="FRS 10-year risk", y="Methylation risk score")

frs0.1_tbl <- rs_compare_data %>%
  filter(frs < 0.1, batch == "fhs_UofMinn") %>%
  mutate(mrs_high_risk=mrs_uofm_holdout > quantile(mrs_uofm_holdout, 0.8)) %>%
  group_by(mrs_high_risk) %>%
  summarise(events=sum(event_10yr),
            n=n()) %>%
  mutate(frac=events / n,
         label=ifelse(mrs_high_risk, 
                      paste0("MRS high-risk\n(n=", n, ")"),
                      paste0("MRS low-risk\n(n=", n, ")")))
frs0.1_plt <- frs0.1_tbl %>%
  ggplot(aes(x=fct_rev(label), y=frac)) +
  geom_bar(stat="identity", width=0.6) +
  labs(x="", y="10-year CVD event fraction",
       subtitle="FHS-UM subgroup: FRS < 0.1") +
  theme(axis.text.x=element_text(angle=20, hjust=0.8),
        axis.text.y=element_text(size=11),
        plot.subtitle=element_text(size=12))

frs0.1_grs0.25q_tbl <- rs_compare_data %>%
  filter(batch == "fhs_UofMinn") %>%
  filter(frs < 0.1 | grs < quantile(grs, 0.25, na.rm=T)) %>%
  mutate(mrs_high_risk=mrs_uofm_holdout > quantile(mrs_uofm_holdout, 0.75)) %>%
  # mutate(mrs_high_risk=cut(mrs_uofm_holdout, 
  #                          breaks=quantile(mrs_uofm_holdout, 
  #                                          seq(0, 1, length.out=5)),
  #                          include.lowest = T)) %>%
  group_by(mrs_high_risk) %>%
  summarise(events=sum(event_10yr),
            n=n()) %>%
  mutate(frac=events / n,
         label=ifelse(mrs_high_risk,
                      paste0("MRS high-risk\n(n=", n, ")"),
                      paste0("MRS low-risk\n(n=", n, ")"))
         )
frs0.1_grs0.25q_plt <- frs0.1_grs0.25q_tbl %>%
  ggplot(aes(x=fct_rev(label), y=frac)) +
  geom_bar(stat="identity", width=0.6) +
  labs(x="", y="10-year CVD event fraction",
       subtitle="FHS-UM subgroup: FRS < 0.1") +
  theme(axis.text.x=element_text(angle=20, hjust=0.8),
        axis.text.y=element_text(size=11),
        plot.subtitle=element_text(size=12))

frs0.2_plt <- rs_compare_data %>%
  filter(frs < 0.2, batch == "fhs_UofMinn") %>%
  mutate(mrs_high_risk=mrs_uofm_holdout > quantile(mrs_uofm_holdout, 0.75)) %>%
  group_by(mrs_high_risk) %>%
  summarise(events=sum(event_10yr),
            n=n()) %>%
  mutate(frac=events / n,
         label=ifelse(mrs_high_risk, 
                      paste0("MRS high-risk\n(n=", n, ")"),
                      paste0("MRS low-risk\n(n=", n, ")"))) %>%
  ggplot(aes(x=fct_rev(label), y=frac)) +
  geom_bar(stat="identity", width=0.6) +
  labs(x="", y="10-year CVD event fraction",
       subtitle="FHS-UM subgroup: FRS < 0.2") +
  theme(axis.text.x=element_text(angle=20, hjust=0.8),
        axis.text.y=element_text(size=11),
        plot.subtitle=element_text(size=12))

# plot_grid(mrs_vs_frs_plt, frs0.1_plt, ncol=2)
```

```{r test-grs}
test_grs <- function(grs_df) {
  grs_df$grs <- scale(grs_df$grs)
  glm((pastEvent | event) ~ grs, data=grs_df, family="binomial") %>%
    tidy() %>%
    filter(term == "grs") %>%
    select(estimate, p.value)
}

grs_res_tbl <- lapply(c(whi="whi", fhs_JHU="fhs_JHU", 
                        fhs_UofMinn="fhs_UofMinn"), 
                      function(b) test_grs(filter(nmd_all, batch == b))) %>%
  bind_rows(.id="Study") %>%
  mutate(Study=study_names_pretty[Study],
         OR_per_SD=round(exp(estimate), 2),
         p=format(p.value, scientific=T, digits=2)) %>%
  select(Study, OR_per_SD, p=p.value)
```

To understand how the present risk score interacts with other established CVD risk metrics, the performance of the MRS was re-evaluated after stratifying individuals by risk scores reflecting either demographic and biochemical features (Framingham Risk Score), or genetic variants (based on Khera et al. 2018). First, the marginal effects of these risk scores were confirmed in each population. The Framingham Risk Score (FRS) was strongly predictive in WHI and FHS, while surprisingly showing no association with CVD incidence in LBC (Supp. Table S4). As the dataset with the largest number of available events, the genetic score was evaluated in WHI, demonstrating a moderate association with CVD (odds ratio per standard deviation = 1.28, p = 1.1e-6).

In pooled Cox models using study-specific baseline hazards and performed using the final four-study MRS, it appeared that the MRS was more effective in those in lower "traditional" risk strata (based on models stratified by FRS categories; Fig. 4a). As a sensitivity analysis, the cohorts were fully stratified into separate models, in which this pattern was visually clear in WHI and FHS-JHU (Supp. Fig. S2). The pattern did not appear in LBC, although we note that the Framingham Risk Score also did not show a "main effect" for incident CVD in this cohort. A similar pattern appeared with respect to genetic risk in WHI (European ancestry participants only based on the formulation of the relevant risk score), in which maximum MRS performance was achieved in the lowest alternative risk stratum. Supplementing these visual comparisons, combined Cox regressions across all cohorts (allowing for different baseline hazards across studies) showed a strong MRS-FRS interaction effect (`r mrs_frs_interaction_effect_per10pct`% reduction in HR for the MRS per 10% increase in FRS; p = `r as.character(mrs_frs_interaction_pval)`), while that for the MRS-GRS interaction did not reach nominal statistical significance (`r mrs_grs_interaction_effect_per_sd`% reduction in HR for the MRS per standard deviation increase in GRS; p = `r as.character(mrs_grs_interaction_pval)`).

To explore the clinical potential of these interactions further, we returned to the initial MRS (trained in 3 datasets with FHS-UM held out). The FHS-UM dataset was filtered to include only participants with lower CVD risk based on the FRS (<10% estimated 10-year risk). Within this lower-risk subset, participants in the upper MRS quintile had more than double the risk of the remainder of the participants: 7% (12/176) of the upper MRS quintile experienced incident events, while 3% (19/701) of the remaining four MRS quintiles experienced incident events. <!--While a similar risk increase was observed when taking the intersection of these low-risk groups (low FRS and low GRS), the sample size was not enough to draw reasonable statistical conclusions (only 6 events in total).-->

FRS could not be calculated in the REGICOR dataset, as not all risk factors were available as continuous values. However, stratified models replicated the observation of greater MRS discrimination in the lowest alternative risk stratum. <!--The estimated odds ratio per standard deviation for the MRS was 6.79 in the youngest individuals (<55 years old) versus 1.29 in the oldest individuals (>65 years old).--> An empirical risk function was generated through logistic regression of MI status on cardiovascular risk factors (age, sex, BMI, diabetes, smoking status, hyperlipidemia (binary), and hypertension). Predicted MI risk using this model was used to stratify subjects into four risk groups, with MRS odds ratios (per standard deviation) of 4.49 in the lowest-risk group versus 1.20 in the highest-risk group. More detailed results from these analyses are shown in Supp. Table S5.

<!--To explore the clinical potential of this methylation-FRS/GRS interaction further, we returned to the initial MRS (trained in 3 datasets with FHS-UM held out). The FHS-UM dataset was filtered to include only participants with low predicted risk based on the FRS (10-year CVD risk <10%). Within this lower-risk subset, participants in the upper MRS quintile had almost triple the risk of the remainder of the participants: 7.4% (13/176) versus 2.6% (18/701).-->

<!--Though continuous risk factor variables were not available for FRS calculation in REGICOR, similar risk strata were calculated based on linear models using demographics and risk factors only. MRS discriminative performance was notably higher in the lowest vs. highest quartiles of this risk metric, agreeing with the training-set results, but did not follow a consistent dose-response relationship.-->

# Discussion

Epigenetic signatures of cardiometabolic diseases and aging in general are being actively explored as biomarkers of disease risk that are potentially modifiable and reveal underlying biological mechanisms. Here, in a novel application of a cross-study ensembling method, we introduce a DNA methylation-based score specific to cardiovascular disease risk. The model performs similarly to one trained on a direct combination of the component datasets, and may perform best in individuals predicted to be at lower risk based on traditional risk factors. 

We opted to use cross-study learning to train our risk model based on the expectation that differences across cohorts (e.g. demographic, behavioral) may contribute to heterogeneity in both the marginal distribution of the CpG features and the conditional distribution of the CVD outcome. Under these conditions, the generalizability of a single-study predictor is often obscured or overstated [@Chang2015; @Zhang2018]. The performance of the CSL model was similar to that of models trained on the merged cohorts with or without batch adjustment via ComBat. This suggests that the assumptions made by these direct combination strategies (i.e. that the heterogeneity structure can be captured by variation in the marginal effects of each CpG site) are met. In practice, this underlying structure is unknown, and we highlight that the CSL was able to produce similar gains in accuracy without making specific assumptions. <!--Using the CSL, we observed improved performance in test sets as compared to the naïve strategy of combining cohorts and training a single-study model. This suggests that up- and down-weighting single-study models offers an advantage compared to the increased combined sample size of the cohort-merging strategy.-->
    
In assessing the stability of the MRS, we observed reasonable reproducibility between technical replicates (ICC=0.85). ICCs for LBC subjects over time were somewhat lower (ICC=0.68), which is to be expected due to not only changes in environment, but also the known epigenetic evolution with age that we observed to be enriched in the components of our score. Furthermore, this value is at the upper end of the range of single-CpG repeatability measurements over time calculated in the combined Lothian Birth Cohorts (1921 and 1936) [@Shah2014]. These ICC values suggest an imperfect but usable reproducibility of the MRS, and an aggregate marker that is fairly robust considering the low replicability that has been observed for individual sites in technical replicates (general median ICC of 0.3 and mode of 0.75 in a "high reliability" cluster) [@Bose2014].

Our observation that different CpGs tended to be selected across studies (Fig. 3) is in agreement with the relative lack of replication seen in prior cardiovascular epigenomic studies [@Fernandez-Sanles2017]. However, the enrichment of the MRS component CpGs for proximity to genes related to cell-cell adhesion (in all subsets except FHS-JHU) is indicative of shared underlying biological mechanisms. As we have previously observed in the WHI and FHS cohorts, it appears that immune activation is central to the prognostic information contained in leukocyte DNA methylation [@Westerman2018]. For example, epigenetic processes have been shown to be involved in the activation and increased adhesion of monocytes in response to environmental insults and metabolic stress, though these have been explored primarily in relation to histone modifications [@Short2017]. Our results provide preliminary support for an attractive model in which a methylation-based score could act as a monitor of cumulative stress in leukocytes and their corresponding activation towards a more atherogenic state.

Existing epigenetic scores have shown varying strengths of association with incident cardiovascular disease. An early investigation examined blood-based methylation in LINE-1 elements, finding strong associations of global hypomethylation with prevalent and incident ischemic heart disease (LINE-1), though additional reports showed opposite associations of methylation at repetitive elements with CVD [@Kim2010]. Guarrera et al. developed a biomarker for MI based on global LINE-1 and ZBTB12 gene methylation that provided a modest net reclassification index improvement (0.23-0.47) compared to traditional risk factors only. Multiple epigenetic aging metrics, though not developed specifically for CVD, have been shown to predict incident CHD, including PhenoAge (odds ratios from 1.02 to 1.08) and GrimAge (hazard ratio = 1.07, adjusted for age and technical factors) [@Levine2018; @Lu2019]. While these associations are statistically significant, they do not represent clinically meaningful improvements in discrimination. Our observed hazard ratio of 1.28 (Basic model in the held-out FHS-UM dataset) indicates that this MRS may be closer to clinical relevance. We note that our component CpG sites overlap strongly with those of these established epigenetic metrics including PhenoAge, suggesting that it captures some of the same biological patterns. However, the mechanistic significance of the specific methylation signals captured by these aging-related metrics, whether as markers of epigenetic regulation breakdown or the work of an "epigenetic maintenance system", is still unclear [@Horvath2013; @Lund2019].

In examining the potential clinical utility of an novel risk score for CVD, it is important to understand to what extent it is redundant with or complementary to existing risk metrics. We first note that the strength of this epigenetic score in adjusted models is lower than that found for traditional risk scores (Supp. Table S4) and some novel biochemical risk measures such as high-sensitivity Troponin I (adjusted HR for global CVD = 3.01) [@Jia2019]. However, analysis of interactions between different risk metrics can be clinically relevant, as demonstrated for example in a recent investigation exploring the interaction between genetic and lifestyle-based risk prediction for dementia [@Licher2019]. Here, we saw a pattern of stronger epigenetic risk associations in individuals whose cardiovascular risk based on traditional metrics (here, the Framingham Risk Score) was low. This pattern replicated in the REGICOR dataset (though FRS could not be directly calculated), with improved MRS discrimination in lower-risk subjects based on an empirical risk function. While these associations are preliminary, they suggest that an epigenetic risk score could help identify higher-risk individuals who otherwise would not have been detected by other metrics. While we did not identify any robust patterns of differential MRS performance in strata based on a genetic cardiovascular risk score, there may have been lower power to detect any such patterns from the outset given the modest discriminatory performance of the GRS in WHI.

Multiple limitations should be acknowledged. While lymphocytes are known to be important in CVD pathogenesis, epigenetic signals have been reported in other CVD-relevant tissues, such as aorta and other vascular tissues [@Fernandez-Sanles2017], that were not examined here. Additionally, the present definition of CVD was chosen to balance specificity of CVD subtypes with sample size, but this balance could be altered to focus on more specific disease subtypes (e.g. myocardial infarction) or a broader definition of CVD (e.g. including heart failure). Finally, while the REGICOR dataset provided an important age- and sex-matched case-control setting for replication of the MRS, this work would benefit from future replication in an independent cohort enabling assessment of incident disease.

In sum, we have developed an epigenetic risk score for cardiovascular disease that provides additional value beyond existing risk measures, and may show improved performance in populations otherwise designated as low-risk. Furthermore, we have shown a novel application of a cross-cohort ensembling method that may provide significant value to future investigations in genomic epidemiology.

# Acknowledgements

We thank all LBC1936 study participants and research team members. The LBC1936 is supported by Age UK (Disconnected Mind program) and the Medical Research Council (MR/M01311/1). Methylation typing in LBC1936 was supported by Centre for Cognitive Ageing and Cognitive Epidemiology (Pilot Fund award), Age UK, The Wellcome Trust Institutional Strategic Support Fund, The University of Edinburgh, and The University of Queensland. LBC1936 work was conducted in the Centre for Cognitive Ageing and Cognitive Epidemiology, which is supported by the Medical Research Council and Biotechnology and Biological Sciences Research Council (MR/K026992/1), and which supported IJD. 


# Reproducibility

Code supporting the analyses described here can be found at https://github.com/kwesterman/meth_cvd. Code and instructions related to the original cross-study learning approach can be found at https://github.com/prpatil/csml_rep. 

# Funding Sources

KW was supported by NIH predoctoral training grant 5T32HL069772-14.

# Disclosures

None

# References