---
title: EWAS for Incident CVD Events
output:
  bookdown::pdf_document2:
    fig_caption: true
    keep_tex: true
    toc: false
    latex_engine: pdflatex
documentclass: bmcart2
bibliography: ../doc/incident_ewas.bib
---

```{r prereqs, include=F}
library(knitr)
opts_chunk$set(echo=F, fig.path="../doc/figures/", cache.path="../cache/incident_ewas/")
suppressMessages(silent <- lapply(c("tidyverse","minfi","survival","doParallel","itertools",
                                    "gridExtra","cowplot"), library, character.only=T))
```

```{r load-data}
# Load methylation data
betas_whi <- readRDS("../int/betas.qc.norm.filt_whi.rds")
betas_fhs <- readRDS("../int/betas.qc.norm.filt_fhs.rds")

# Load metadata
metaData <- readRDS("../int/metaData.rds")

# Load estimated cell counts
estCellCounts_whi <- readRDS("../int/estCellCounts_whi.rds")
estCellCounts_fhs <- readRDS("../int/estCellCounts_fhs.rds")

# Load CPACOR principal component adjustment factors
load("../int/CPACOR_whi.RData")
CP_PCs_whi <- CP_PCs
load("../int/CPACOR_fhs.RData")
CP_PCs_fhs <- CP_PCs

# Load PCA results
load("../int/PCA.fit_whi.RData")
PCs_whi <- PCs
load("../int/PCA.fit_fhs.RData")
PCs_fhs <- PCs

# Load Illumina 450k annotation
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
anno450k <- data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19), stringsAsFactors=F)
```

```{r clean-data, warning=F}
estCellCounts <- bind_rows(estCellCounts_whi, estCellCounts_fhs)
CP_PCs <- bind_rows(CP_PCs_whi, CP_PCs_fhs)
PCs <- bind_rows(PCs_whi, PCs_fhs)

### Create master covariate/event data frame
nonMethData <- Reduce(function(x,y) inner_join(x, y, by="sampleKey"), 
                          list(metaData, estCellCounts, CP_PCs, PCs)) %>%
  distinct(subjID, .keep_all=T)  # Hacky...better way of deciding which biological replicates to keep?
nonMethData <- replace_na(nonMethData, 
                          list(bmi=median(nonMethData$bmi, na.rm=T),
                               smk_now=0, smk_py=0, ht_med=0, lipid_med=0, dm_med=0)) %>%
  mutate(diabetes=dm_med | glu>125) %>%
  mutate_at(c("chol","ldl","hdl","tg","sbp","glu","hscrp"), log10)
nmd_whi <- filter(nonMethData, study=="whi")
nmd_fhs <- filter(nonMethData, study=="fhs")

betas_whi <- betas_whi[,match(nonMethData$sampleKey[nonMethData$study=="whi"], colnames(betas_whi))]
betas_fhs <- betas_fhs[,match(nonMethData$sampleKey[nonMethData$study=="fhs"], colnames(betas_fhs))]
stopifnot(all(colnames(betas_whi)==nonMethData$sampleKey[nonMethData$study=="whi"]),
          all(colnames(betas_fhs)==nonMethData$sampleKey[nonMethData$study=="fhs"]))  # Ensure identical number and order of samples for methylation and covariate data
```

# Introduction

Genetic approaches to cardiovascular disease (CVD) research have led to important breakthroughs in mechanistic understanding and therapeutic strategies. However, the mechanisms for gene variant-disease relationships are often difficult to determine, and their effects may often be mediated by epigenetic regulation [@Bonder2016]. DNA methylation is one such mechanism that can integrate both genetic variation and environmental exposures and potentially reflect or drive their effects on CVD outcomes [@Ordovas2010].

A series of recent epigenome-wide association studies (EWAS) have examined relationships between DNA methylation at cytosine-phosphate-guanine (CpG) sites and various CVD phenotypes, including prior myocardial infarction [@Rask-Andersen2016], acute coronary syndrome [@Li2017], and atherosclerosis [@Nakatochi2017]. These discoveries may reveal important mechanistic insights, but are susceptible to reverse causation. Mendelian randomization approaches can resolve some of this ambiguity, but are dependent on the existence of appropriate genetic instrumental variables (methylation-quantitative trait loci) for the CpGs of interest. Furthermore, those that have been performed often suggest that reverse causation (methylation being influenced by the phenotype of interest) is more common [@Dekkers2016;@Wahl2017].

One approach to this problem is to examine epigenetic associations with cardiovascular risk factors. Multiple investigations have explored these relationships genome-wide [@Pfeiffer2015;@Irvin2014], and have even uncovered prognostic CpG sites for incident coronary heart disease in the process [@Hedman2017]. A few studies looking directly at incident CVD as a binary variable have found relationships with global DNA methylation (as approximated by LINE-1 methylation levels) and with a specific cluster of CpG sites [@Baccarelli2010;@Guarrera2015].

Most epigenetic investigations of CVD to-date have been performed in blood, based on the importance of immune-related processes in its pathogenesis [@Back2015] as well as its relative ease of collection in comparison to other relevant tissues (e.g. endothelium, myocardium). In addition, several groups have shown a moderate ability of blood-based methylomic signatures to reflect those in disease-relevant tissues [@Bacos2016;@Huang2016].

To address the problem of reverse causation while incorporating time-to-event information, we performed a Cox model-based EWAS for incident CVD in a series of two cohorts. We used sensitivity analyses to demonstrate that demographic factors are relevant, but not dominant, in explaining the variation in results across cohorts. Furthermore, we incorporated existing genetic, epigenetic, and transcriptomic datasets to explore the potential mechanistic relevance of different cell types and transcription factors to CVD risk, both within blood and in other cell types of interest. Finally, we assessed relationships between methylation at relevant CpG sites and traditional cardiovascular risk factors.

# Results

## Genome-wide associations between DNA methylation and incident CVD events

```{r qc, warning=F}
make_variation_plot <- function(cohort) {
  nmd <- filter(nonMethData, study==cohort)
  
  wbcpca.fit <- prcomp(select(nmd, CD8T, CD4T, NK, Bcell, Mono, Gran), scale.=T)
  nmd$WBC_PC1 <- wbcpca.fit$x[,"PC1"]
  nmd$WBC_PC2 <- wbcpca.fit$x[,"PC2"]
  
  covs <- c("sex","age","race","bmi","smk_now","smk_py","WBC_PC1","WBC_PC2",
            "plate","center","dnaPull","sentrixRow","sentrixCol","event","pastEvent")
  comparisonDF <- expand.grid(pc=c(paste0("PC", 1:6),"cpPC1","cpPC2","event"), covar=covs)
  
  comparisonDF$negLogP <- apply(comparisonDF, 1, function(row) {
    pc <- row["pc"]
    covar <- row["covar"]
    nlp <- tryCatch({
      lm.fit <- lm(as.formula(paste0(pc, "~", covar)), data=nmd)
      pVal <- anova(lm.fit)[covar,"Pr(>F)"]
      -log10(pVal)
    }, error=function(e) 0)
    if(is.na(nlp)) 0 else min(nlp, 20)
  })
  
  ggplot(comparisonDF, aes(x=pc, y=forcats::fct_rev(covar), fill=negLogP)) +
    geom_tile(color="white") +
    scale_fill_gradient2(low='#0000FF',mid='#FFFFFF',high='#FF0000') +
    scale_x_discrete(position="top") +
    labs(title=toupper(cohort),
         caption="Note: -log(p-values) capped at 20 to aid visualization.") +
    theme(axis.title=element_blank())
}

whiVarPlt <- make_variation_plot("whi")
fhsVarPlt <- make_variation_plot("fhs")
```
```{r ewas-prep}
myTry <- function(expr, CpG) {
  # Captures model failures and returns successful result or vector of NAs
  tryCatch(expr,
           error = function(e) {print(e); return(c(CpG, rep(NA, 3)))})
}

run_coxModel <- function(probeData, covarData, model_spec) {
  # Given a row of the M-value matrix (corresponding to a CpG site), bind that methylation
  # data to the covariate data and run Cox proportional hazards regression
  CpG <- rownames(probeData)
  probeData <- as.vector(probeData)
  outlierTF <- probeData<quantile(probeData,0.25)-5*IQR(probeData) | 
    probeData>quantile(probeData,0.75)+5*IQR(probeData)
  modelData <- cbind(covarData, meth=as.numeric(probeData),
                     survObj=Surv(time=covarData$time, event=covarData$event))
  myTry({
    cox.fit <- coxph(as.formula(model_spec), data=modelData, subset=!outlierTF)
    c(CpG=CpG, summary(cox.fit)$coef['meth',c('coef','z','Pr(>|z|)')])
  }, CpG)
}
```
```{r ewas-discovery, cache=1}
ewasModel_whi <- paste0("survObj~meth+age+race+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+dnaPull")

cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)

whiResList <- foreach(meth=iter(betas_whi, by="row"), .packages="survival") %dopar%
  run_coxModel(meth, nonMethData[nonMethData$study=="whi",], ewasModel_whi)

stopCluster(cl)

whiRes <- do.call(rbind, whiResList) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
```
```{r ewas-replication, cache=1}
ewasModel_fhs <- paste0("survObj~meth+sex+age+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+center+",
                        paste0("cpPC",1:7,collapse="+"))

cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)

fhsResList <- foreach(meth=iter(betas_fhs, by="row"), .packages="survival") %dopar%
  run_coxModel(meth, nonMethData[nonMethData$study=="fhs",], ewasModel_fhs)

stopCluster(cl)

fhsRes <- do.call(rbind, fhsResList) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
```
```{r ewas-inflation}
make_qqplot <- function(pVec, plotTitle="Title") {
  pVec <- pVec[!is.na(pVec)]
  qqplot(-log10(1:length(pVec)/length(pVec)), -log10(pVec), pch=".", main=plotTitle, xlab="Expected (-logP)", ylab="Observed (-logP)")
  abline(0,1,col="red")
}
gControl <- function(pVals) {
# See van Iterson 2017 methods and/or Lehne 2015 code for details on genomic control for EWAS
# Below is modeled after Lehne 2015
lambda <- median(qchisq(pVals, df=1, lower.tail=F), na.rm=T)/qchisq(0.5, df=1)
round(lambda, 2)
}

# whiQQ <- make_qqplot(whiRes$p, plotTitle=paste0("lambda = ", gControl(whiRes$p)))
# fhsQQ <- make_qqplot(fhsRes$p, plotTitle=paste0("lambda = ", gControl(fhsRes$p)))
```
```{r hachiya, cache=1, dependson=c("ewas-discovery","ewas-replication")}
hachiya450k <- readRDS("../int/hachiya450k.rds")
hachiyaHighRI <- filter(hachiya450k, Beta_RI.mono>30 | Beta_RI.cd4t>30 | Beta_RI.neu>30)$Name

betaRIs_whi <- apply(betas_whi, 1, function(betas) diff(quantile(betas, c(0.05,0.95))))
betaRIs_fhs <- apply(betas_fhs, 1, function(betas) diff(quantile(betas, c(0.05,0.95))))

allRIs <- tibble(cpg=c(names(betaRIs_whi), names(betaRIs_fhs)),
                 RI=c(betaRIs_whi, betaRIs_fhs),
                 study=rep(c("WHI","FHS"), times=c(length(betaRIs_whi),length(betaRIs_fhs))),
                 "Hachiya variable site"=cpg %in% hachiyaHighRI)
hachiyaPlt <- ggplot(allRIs, aes(x=study, y=RI, fill=`Hachiya variable site`)) +
  geom_boxplot() +
  ylab("Reference Interval (diff. between 5th and 95th %ile)") +
  theme(axis.title.x=element_blank())

whiResFull <- whiRes
whiRes <- whiResFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
fhsResFull <- fhsRes
fhsRes <- fhsResFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
```

The epigenome-wide association analysis was performed using two datasets: the Women's Health Initiative Study (WHI) for discovery, and the Framingham Heart Study Offspring Cohort (FHS) for validation. Subject characteristics are shown in Table \@ref(tab:population-description). The objective of our initial analysis was to identify specific CpG sites whose methylation levels in blood at a given baseline timepoint predict incident cardiovascular events. For each CpG site passing quality control (`r nrow(betas_whi)` in WHI and `r nrow(betas_fhs)` in FHS), a Cox proportional hazards model was used to predict time-to-event from methylation beta values while adjusting for relevant biological and technical covariates.

```{r population-description}
popData <- nonMethData %>%
  dplyr::select(study, sex, age, race, bmi, smk_now, pastEvent, event) %>%
  group_by(study) %>%
  summarise(n=n(),
            perc_female=paste(round(sum(sex=="F")/n*100), " %"),
            age_range=paste0(median(age), " (", paste(range(age),collapse="-"), ")"),
            mixed_race=ifelse(length(unique(race))>1, "Yes", "No"),
            bmi_range=paste0(round(median(bmi),1), " (", paste(round(range(bmi),1),collapse="-"), ")"),
            perc_smoke=paste0(round(sum(smk_now)/n*100), " %"),
            n_pastEvent=sum(pastEvent),
            n_event=sum(event)) %>%
  setNames(c("Dataset","Sample Size","% female","Age","Multiple races","BMI","% smoke",
             "# prior CVD events","# incident CVD events")) %>%
  t() %>%
  data.frame() %>%
  setNames(c("FHS","WHI"))
kable(popData[-1,], caption="Population description.", booktabs=T)
```

Genomic inflation was limited in both datasets across the full set of CpG sites (see Supp. Fig. \@ref(fig:supp-qq-plots)), suggesting adequate correction for substructure and confounding. However, to increase power, a subset of `r length(hachiyaHighRI)` highly variable sites were taken from a prior analysis of methylation variability in blood cell subtypes [@Hachiya2017] and used during the initial analysis. Though this dataset was gathered in Asian subjects, the variability patterns are highly consistent with those seen in the datasets analyzed here (Supp. Fig. \@ref(fig:supp-hachiya-plot)).

```{r ewas-discovery-investigation, warning=F}
qqmanInput_whi <- whiRes %>%
  inner_join(select(anno450k, Name, chr, pos), by=c("CpG"="Name")) %>%
  mutate(CHR=as.numeric(gsub("chr","",chr)), BP=as.numeric(pos), P=as.numeric(p)) %>%
  na.omit()
```
```{r top-hits}
bonferroniThreshold <- 0.05/nrow(whiRes)
whiToFhsReplication <- whiRes %>%
  filter(p<bonferroniThreshold, !is.na(p)) %>%
  left_join(fhsRes, by="CpG", suffix=c(".whi",".fhs"))
bonferroniTbl <- whiToFhsReplication %>%
  select(-contains("z"), -contains("fdr"))
# kable(bonferroniTbl, caption="Bonferroni-significant CpGs in the discovery set")
fdrTbl <- whiRes %>%
  filter(fdr<0.05) %>%
  mutate(direction=ifelse(sign(coef)==1, "+", "-"),
         p=as.character(signif(p,3))) %>%
  inner_join(select(anno450k, Name, chr, UCSC_RefGene_Group, UCSC_RefGene_Name), by=c("CpG"="Name")) %>%
  select(CpG, chr, direction, p, UCSC_RefGene_Group, UCSC_RefGene_Name) %>%
  mutate_at(c("UCSC_RefGene_Group","UCSC_RefGene_Name"), function(n) gsub(";.*","",n)) %>%
  setNames(c("CpG","Chromosome","Association","P-value","Location","Annotated gene"))
kable(fdrTbl, caption="CpGs with FDR<0.05 in the discovery set.", booktabs=T)
```
```{r suggestive-overlap}
# bothNominalSites <- na.omit(intersect(whiRes$CpG[whiRes$p<0.01], fhsRes$CpG[fhsRes$p<0.01]))
bothSuggestiveSites <- na.omit(intersect(whiRes$CpG[whiRes$fdr<0.25], fhsRes$CpG[fhsRes$fdr<0.25]))
eitherSuggestiveSites <- na.omit(union(whiRes$CpG[whiRes$fdr<0.25], fhsRes$CpG[fhsRes$fdr<0.25]))

whiCoefs <- setNames(whiRes$coef, whiRes$CpG)
fhsCoefs <- setNames(fhsRes$coef, fhsRes$CpG)
```
```{r past-event-strat, cache=1}
fhsPEStratModel <- paste0("survObj~meth+sex+age+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+center+",
                           paste0("cpPC",1:7,collapse="+"))
cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)
fhsResListPE <- foreach(meth=iter(betas_fhs, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_fhs$pastEvent==T], nmd_fhs[nmd_fhs$pastEvent==T,], fhsPEStratModel)
fhsResPE <- do.call(rbind, fhsResListPE) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_fhs)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
fhsResListNoPE <- foreach(meth=iter(betas_fhs, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_fhs$pastEvent==F], nmd_fhs[nmd_fhs$pastEvent==F,], fhsPEStratModel)
fhsResNoPE <- do.call(rbind, fhsResListNoPE) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_fhs)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
stopCluster(cl)
```
```{r site-investigation}
suggestiveOrder <- order(abs(whiCoefs[bothSuggestiveSites]+fhsCoefs[bothSuggestiveSites])/2, decreasing=T)
suggestiveMyAnnot <- anno450k[match(bothSuggestiveSites, anno450k$Name),] %>%
  dplyr::select(Name, chr, UCSC_RefGene_Name, UCSC_RefGene_Group) %>%
  dplyr::slice(suggestiveOrder) %>%
  mutate(Gene=unlist(lapply(strsplit(UCSC_RefGene_Name, split=";"), 
                            function(x) paste(unique(x), collapse=";"))),
         Location=unlist(lapply(strsplit(UCSC_RefGene_Group, split=";"), 
                            function(x) paste(unique(x), collapse=";")))) %>%
  dplyr::rename(CpG=Name, Chr=chr) %>%
  mutate(Association=ifelse(sign(whiCoefs[CpG])==1, "+", "-")) %>%
         # Tissue_GTEx=c("Lymph;Fibroblast;SkM",
         #               "Whole blood",
         #               "Brain",
         #               "Lymph;Spleen",
         #               "Lung;Spleen;Adrenal",
         #               "Artery",
         #               "Lymph;Fibro",
         #               "Artery;Testis",
         #               "Skin;~Artery",
         #               "Lymph;Fibro"),
         # My_notes=c("Glycolysis; uracil glycosylase activity",
         #            "GTP-related functions",
         #            "AD-related, interacts w/ APP",
         #            "B-lymph. surface molecule",
         #            "RCT; consistently found in lipid EWAS",
         #            "AD-related, interacts w/ APP & MHC1 molecules",
         #            "Disulfide isomerase",
         #            "Cytoskeleton",
         #            "p53/NFkB regulator",
         #            "tRNA synthetase")) %>%
  select(CpG, Association, Chr, Location, Gene)
kable(suggestiveMyAnnot, caption="8 CpGs found at FDR<0.25 in both datasets.", booktabs=T)

boxplotData <- cbind(nonMethData[,c("study","event")], 
                     t(cbind(betas_fhs[bothSuggestiveSites,], betas_whi[bothSuggestiveSites,]))) %>%
  gather(key=CpG, value=Beta, -study, -event)
dmpBoxplot <- ggplot(boxplotData, aes(x=CpG, y=Beta, fill=event)) +
  geom_boxplot() +
  # ggtitle("8 CpGs found at FDR<0.25 in both datasets") +
  theme(axis.text.x=element_text(angle=20), axis.title.x=element_blank()) +
  facet_wrap(~study, nrow=2)
```

Sites passing a false discovery rate threshold of 0.05 are shown in Table \@ref(tab:top-hits) (a Manhattan plot of the full results can be found in Supp. Fig. \@ref(fig:supp-manhattan)). `r nrow(bonferroniTbl)` CpG site was found to pass a Bonferroni threshold of p<`r as.character(signif(bonferroniThreshold,2))` in the discovery cohort, though it replicated only marginally in the validation cohort (p=`r round(fhsRes$p[fhsRes$CpG==whiRes$CpG[1]],2)`). Despite this lack of validation, there was consistency in the direction of coefficients between cohorts. Of `r length(bothSuggestiveSites)` sites reaching a suggestive FDR<0.25 in both datasets, `r round(sum(sign(whiCoefs[bothSuggestiveSites])==sign(fhsCoefs[bothSuggestiveSites]))/length(bothSuggestiveSites)*100)`% show consistent direction of effect, and will be referred to as differentially methylated positions (DMPs; Table \@ref(tab:site-investigation)).

The effect of confounding by past cardiovascular events in FHS was minimized in the analysis setup by using FHS as a replication rather than discovery set. However, to further address this possibility, EWAS analyses stratified by past event status were conducted. Of the 10 suggestive sites overlapping at FDR<0.25, both strata showed perfect consistency of regression coefficient signs with those from WHI. Though power to identify strong signals was reduced by the stratification, `r sum(fhsResNoPE$p[match(bothSuggestiveSites,fhsResNoPE$CpG)]<0.05)` of 8 DMPs retained nominal significance (p<0.05) in the group without past events, compared to `r sum(fhsResPE$p[match(bothSuggestiveSites,fhsResPE$CpG)]<0.05)` sites in the group with past events. These results indicate that the moderate agreement between cohorts was not being primarily driven by history of CVD.

```{r genome-distribution}
summarise_annotation <- function(group, df) {
  # Takes in a set of annotations and returns the proportions of a given field (e.g. gene name)
  df %>%
    select(one_of("Name", group)) %>%
    separate_rows(group, sep=";") %>%
    distinct() %>%
    filter_at(group, all_vars(.!="")) %>%
    group_by_at(group) %>%
    summarise(n=n()) %>%
    mutate(frac=n/sum(n))
}

annot_plot <- function(cpgSet, bgCpgSet, group_var) {
  # Given a specific field, plot proportion of CpGs in a specific subset next to those of the full set
  my_groups <- summarise_annotation(group_var, filter(anno450k, Name %in% cpgSet))
  all_groups <- summarise_annotation(group_var, filter(anno450k, Name %in% bgCpgSet))
  merge_groups <- inner_join(my_groups, all_groups, by=group_var, suffix=c(".subset",".all")) %>%
    gather(key=group, value=frac, contains("frac")) %>%
    mutate(group=factor(group, levels=c("frac.all","frac.subset"), labels=c("All CpGs","EWAS hits")))
  ggplot(merge_groups, aes_string(x=group_var, y="frac", fill="group")) +
    geom_bar(stat="identity", position="dodge") +
    theme(axis.title=element_blank(), axis.text.x=element_text(angle=30, hjust=1))
}

geneGroupEnrichmentPlot <- annot_plot(eitherSuggestiveSites, hachiyaHighRI, "UCSC_RefGene_Group")
cpgIslandEnrichmentPlot <- annot_plot(eitherSuggestiveSites, hachiyaHighRI, "Relation_to_Island")
# annot_plot(topCpGs, "Regulatory_Feature_Group")

calc_enrichment_p <- function(group, lowerTail=F) {
  rtiSummarySuggestive <- summarise_annotation("Relation_to_Island", filter(anno450k, Name %in% eitherSuggestiveSites))
  rtiSummaryHachiya <- summarise_annotation("Relation_to_Island", filter(anno450k, Name %in% hachiyaHighRI))
  n_hitsInGroup <- sum(rtiSummarySuggestive$n[rtiSummarySuggestive$Relation_to_Island %in% group])
  n_availInGroup <- sum(rtiSummaryHachiya$n[rtiSummaryHachiya$Relation_to_Island %in% group])
  phyper(n_hitsInGroup, n_availInGroup, length(hachiyaHighRI)-n_availInGroup, length(eitherSuggestiveSites), lower.tail=lowerTail)
}

pIsland <- calc_enrichment_p("Island")
pShore <- calc_enrichment_p(c("N_Shore","S_Shore"))
pOpenSea <- calc_enrichment_p("OpenSea", lowerTail=T)
```

Because only `r length(bothSuggestiveSites)` sites overlapped at FDR<0.25, the union rather than intersection of these sites (`r length(eitherSuggestiveSites)` CpGs) was used for further investigation of the characteristics of the set. These sites showed no major enrichment for gene-related genomic locations (as compared to a background of only the highly variable subset), but did show a moderate enrichment for presence in CpG islands (p=`r as.character(signif(pIsland,2))`; hypergeometric test) and shores (p=`r as.character(signif(pShore,2))`) 
<!--at the expense of those in the "open sea" (p=`r as.character(signif(pOpenSea,2))` for depletion).-->

```{r gsea, include=F}
gsaTbl <- missMethyl::gometh(as.vector(eitherSuggestiveSites), all.cpg=hachiyaHighRI, collection="GO")
gsaTopResTbl <- arrange(filter(gsaTbl, FDR<0.05), FDR)
```

A gene set enrichment analysis of genes annotated to the same set of DMPs showed enrichment for biological processes related to developmental processes (embryonic development, cell fate commitment, and embryonic morphogenesis), cell adhesion, and transcriptional regulation (see Supp. Table \@ref(tab:supp-gsea-table)).

## Demographic stratification partially explains the observed heterogeneity between cohorts

```{r sex-strat, cache=1}
fhsSexStratModel <- paste0("survObj~meth+age+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+center+",
                           paste0("cpPC",1:7,collapse="+"))
cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)
fhsResListMale <- foreach(meth=iter(betas_fhs, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_fhs$sex=="M"], nmd_fhs[nmd_fhs$sex=="M",], fhsSexStratModel)
fhsResMale <- do.call(rbind, fhsResListMale) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_fhs)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
fhsResListFemale <- foreach(meth=iter(betas_fhs, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_fhs$sex=="F"], nmd_fhs[nmd_fhs$sex=="F",], fhsSexStratModel)
fhsResFemale <- do.call(rbind, fhsResListFemale) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_fhs)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
stopCluster(cl)
```
```{r race-strat, cache=1}
whiRaceStratModel <- "survObj~meth+age+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+dnaPull"
cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)
whiResListWhite <- foreach(meth=iter(betas_whi, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_whi$race=="white"], nmd_whi[nmd_whi$race=="white",], whiRaceStratModel)
whiResWhite <- do.call(rbind, whiResListWhite) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_whi)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
whiResListBlack <- foreach(meth=iter(betas_whi, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_whi$race=="black"], nmd_whi[nmd_whi$race=="black",], whiRaceStratModel)
whiResBlack <- do.call(rbind, whiResListBlack) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_whi)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
whiResListHispanic <- foreach(meth=iter(betas_whi, by="row"), .packages="survival") %dopar%
  run_coxModel(meth[nmd_whi$race=="hispanic"], nmd_whi[nmd_whi$race=="hispanic",], whiRaceStratModel)
whiResHispanic <- do.call(rbind, whiResListHispanic) %>%
  data.frame(stringsAsFactors=F, check.names=F) %>%
  mutate(CpG=rownames(betas_whi)) %>%
  dplyr::rename(p=`Pr(>|z|)`) %>%
  mutate_at(c("coef","z","p"), as.numeric) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)
stopCluster(cl)
```

Sex and race were adjusted for in the above analyses in order to remove their direct associations with incident CVD events, but interactions with methylation were not included that would allow loci to have different associations between demographics. To evaluate the potential role of this demographic heterogeneity in explaining the level of agreement between cohorts, stratified analyses were undertaken (separates sexes in FHS, and separate races in WHI).

```{r strat-analyses, warning=F, cache=1}
fhsResMaleFull <- fhsResMale
fhsResMale <- fhsResMaleFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
fhsResFemaleFull <- fhsResFemale
fhsResFemale <- fhsResFemaleFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
whiResWhiteFull <- whiResWhite
whiResWhite <- whiResWhiteFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
whiResBlackFull <- whiResBlack
whiResBlack <- whiResBlackFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
whiResHispanicFull <- whiResHispanic
whiResHispanic <- whiResHispanicFull %>%
  filter(CpG %in% hachiyaHighRI) %>%
  mutate(fdr=p.adjust(p, method="BH"))
# fhsResPEFull <- fhsResPE
# fhsResPE <- fhsResPEFull %>%
#   filter(CpG %in% hachiyaHighRI) %>%
#   mutate(fdr=p.adjust(p, method="BH"))
# fhsResNoPEFull <- fhsResNoPE
# fhsResNoPE <- fhsResNoPEFull %>%
#   filter(CpG %in% hachiyaHighRI) %>%
#   mutate(fdr=p.adjust(p, method="BH"))

compare_results <- function(resMain, resCompare, comparison) {
  compareRanks <- 1:sum(resMain$fdr<0.25)
  corTest <- cor.test(resMain$coef[compareRanks], 
           resCompare[match(resMain$CpG[compareRanks],resCompare$CpG),"coef"])
  bothSuggestive <- resCompare$CpG[resCompare$CpG %in% resMain$CpG[resMain$fdr<0.25] & resCompare$fdr<0.25]
  # bothNominal <- resCompare$CpG[resCompare$CpG %in% resMain$CpG[resMain$p<0.01] & resCompare$p<0.01]
  list(df=tibble(comparison=comparison, corr=corTest$estimate, corrSig=corTest$p.value, 
                 suggestiveOverlap=length(bothSuggestive)),
       commonProbes=bothSuggestive)
}

compareList <- list(whi2fhs=compare_results(whiRes, fhsRes, "whi2fhs"),
                    whi2fhsMale=compare_results(whiRes, fhsResMale, "whi2fhsMale"),
                    whi2fhsFemale=compare_results(whiRes, fhsResFemale, "whi2fhsFemale"),
                    # whi2fhsPE=compare_results(whiRes, fhsResPE, "whi2fhsPastEvents"),
                    # whi2fhsNoPE=compare_results(whiRes, fhsResNoPE, "whi2fhsNoPastEvents"),
                    fhs2whi=compare_results(fhsRes, whiRes, "fhs2whi"),
                    fhs2whiWhite=compare_results(fhsRes, whiResWhite, "fhs2whiWhite"),
                    fhs2whiBlack=compare_results(fhsRes, whiResBlack, "fhs2whiBlack"),
                    fhs2whiHispanic=compare_results(fhsRes, whiResHispanic, "fhs2whiHispanic"))

compareDF <- bind_rows(map(compareList, "df")) %>%
  mutate(group=rep(c("Full WHI to stratified FHS","Full FHS to stratified WHI"), times=c(3,4)),
         group=factor(group, levels=unique(group)),
         stratum=c("Full","Male","Female","Full","White","Black","Hispanic"),
         stratum=factor(stratum, levels=unique(stratum)))

# ggplot(comparisonsDF, aes(x=compareDirection, y=corr)) +
# geom_bar(aes(group=comparison), stat="identity", position="dodge", width=0.9)
stratPlt <- ggplot(compareDF, aes(x=stratum, y=corr)) +
  geom_bar(stat="identity") +
  labs(y="Correlation") +
  theme(axis.title.x=element_blank()) +
  facet_wrap(~group, scales="free_x")

stratOverlapPlt <- ggplot(compareDF, aes(x=comparison, y=suggestiveOverlap)) +
  geom_bar(stat="identity", width=0.9) +
  ggtitle("Size of overlap between probes with FDR<0.25 in each cohort/subset combination")

gsaStrat <- map(list(fhsRes, fhsResMale, fhsResFemale, whiRes, whiResWhite, whiResBlack, whiResHispanic),
                function(stratRes) tryCatch(missMethyl::gometh(stratRes$CpG[stratRes$fdr<0.25], 
                                                               all.cpg=hachiyaHighRI,
                                                               collection="GO"),
                                            error=function(e) NA))
# map(gsaStrat[!is.na(gsaStrat)], function(gsaResult)  head(arrange(gsaResult, P.DE)))
# gsaCompStrat <- map(compareList, function(comp) gometh(comp$commonProbes, collection="GO"))
# map(gsaCompStrat, function(gsaResult) head(arrange(gsaResult, P.DE)))

## IMPLEMENT proportion test: prop.test(c(counts), c(trials))
```

```{r strat-plt, out.width="70%", fig.cap="Pearson correlations between EWAS regression coefficients in one dataset with FDR<0.25 and those of the same CpGs in subsets of the other."}
stratPlt
```

Fig. \@ref(fig:strat-plt) shows the correlations between regression coefficients from a main cohort (w/ FDR<0.25) and those of the corresponding sites from the stratified groups. In this dataset, similarity of demographic factors did associate with similarity of regression coefficients: Caucasians in WHI displayed more similar coefficients to FHS and females in FHS displayed more similar coefficients to WHI. However, these comparisons of more similar demographic groups did not uncover an increased number of replicating loci at FDR<0.25. 

A clear lack of correlation with FHS was observed in black subjects in comparison to white or hispanic subjects from WHI. Although females from FHS did show a slightly higher correlation with WHI, this difference was not large, and neither stratum correlated as well as the full FHS dataset. Notably, however, after stratification of the FHS dataset by sex, the single CpG site passing the Bonferroni threshold in WHI (upstream of the TSS of VKORC1) showed a nominal association in women (p = `r round(fhsResFemale$p[fhsResFemale$CpG==whiRes$CpG[1]],3)`) but none in men. 

## Region-based approach detects additional associations

In order to expand the site-by-site analysis and take advantage of shared information between proximal CpG sites, the Comb-p algorithm was used to identify differentially methylated regions (DMRs). Comb-p seeks regions enriched for low p-values while accounting for autocorrelation based on genomic distance. It was used separately on WHI and FHS (with p-values from the above EWAS used as input), and uses the full set of loci (rather than the variance-based subset) for maximum genome coverage.

```{r combp-prep}
res_to_bed <- function(resDF) {
  resDF %>%
    inner_join(anno450k, by=c("CpG"="Name")) %>%
    mutate(chrom=chr, chromStart=pos-1, chromEnd=pos, name=CpG) %>%  # Start at provided genomic coordinate - 1 (based on a script from Github)
    dplyr::select(chrom, chromStart, chromEnd, p, name) %>%
    filter(!is.na(p)) %>%
    arrange(chrom, chromStart)  # Order by chromosome/position
}

write_tsv(res_to_bed(whiResFull), "../int/whiRes.bed")
write_tsv(res_to_bed(fhsResFull), "../int/fhsRes.bed")
```
```{r combp-results, message=F}
combp_res_whi <- read_tsv("../int/regions.sig_whi.bed", col_names=F) %>%
  setNames(c("chr","start","end","lowest_p","num_cpgs","p_region_slk","p_adj_region_sidak")) %>%
  mutate(p_adj_region=signif(pmin(p_region_slk*(nrow(whiResFull)/num_cpgs), 1), 3)) %>%
  arrange(p_adj_region) %>%
  filter(p_region_slk<0.05)
combp_ranges_whi <- GRanges(combp_res_whi$chr, IRanges(combp_res_whi$start, combp_res_whi$end))
combp_res_fhs <- read_tsv("../int/regions.sig_fhs.bed", col_names=F) %>%
  setNames(c("chr","start","end","lowest_p","num_cpgs","p_region_slk","p_adj_region_sidak")) %>%
  mutate(p_adj_region=signif(pmin(p_region_slk*(nrow(fhsResFull)/num_cpgs), 1), 3)) %>%
  arrange(p_adj_region) %>%
  filter(p_region_slk<0.05)
combp_ranges_fhs <- GRanges(combp_res_fhs$chr, IRanges(combp_res_fhs$start, combp_res_fhs$end))

overlap_ranges <- findOverlaps(combp_ranges_whi, combp_ranges_fhs)
combpResTbl <- combp_res_whi %>%
  dplyr::slice(from(overlap_ranges)) %>%
  dplyr::select(chr, start, end, p_region_slk, p_adj_region) %>%
  cbind(p_region_slk_fhs=combp_res_fhs$p_region_slk[to(overlap_ranges)],
        p_adj_region_fhs=combp_res_fhs$p_adj_region[to(overlap_ranges)]) %>%
  filter(p_adj_region<0.05,
         p_region_slk_fhs<0.05)

find_genes_byRegion <- function(locDF) {
  apply(locDF, 1, function(row) {
    regionAnnot <- filter(anno450k, chr==row["chr"], 
                          pos>as.numeric(row["start"]), pos<=as.numeric(row["end"]))
    gsub(";.*", "", regionAnnot$UCSC_RefGene_Name[1])
  })
}

combpResTbl <- combpResTbl %>%
  mutate(annot_gene=find_genes_byRegion(.)) %>%
  mutate(loc=paste0(chr, ":", start, "-", end),
         gene_loc=c("CpG island at TSS","CpG island in body","Body","CpG island"),
         num_cpgs=c(12,8,3,5)) %>%
  mutate_at(vars(contains("p_")), function(p) as.character(signif(p,3))) %>%
  select(chr, start, end, loc, num_cpgs, annot_gene, gene_loc, p_region_slk, p_adj_region, p_region_slk_fhs)

kable(setNames(select(combpResTbl, loc, num_cpgs, annot_gene, gene_loc, 
                      p_region_slk, p_adj_region, p_region_slk_fhs), 
               c("Location","# CpGs","Annotated gene","Genomic region","P.whi","Padj.whi","P.fhs")),
      caption="Comb-p regions with multiple test-corrected p<0.05 in WHI and uncorrected p<0.05 in FHS",
      booktabs=T)

whiDmrs <- combp_res_whi[1:20,] %>% mutate(annot_gene=find_genes_byRegion(.)) %>% select(-lowest_p)
fhsDmrs <- combp_res_fhs[1:20,] %>% mutate(annot_gene=find_genes_byRegion(.)) %>% select(-lowest_p)
```

In contrast to the site-based approach, `r nrow(combpResTbl)` regions were identified that passed a Bonferroni correction in the discovery set and showed at least nominal significance in validation (Table \@ref(tab:combp-results)). Three of these regions are annotated to known genes: ApoB (N shore of island in gene body), ZNF274 (CpG island in gene body) and SLC9A1 (gene body). Fig. \@ref(fig:combp-plots) shows EWAS associations within these regions, with and without adjustment for traditional cardiovascular risk factors (LDL-C, HDL-C, Tg, Sys. BP, hsCRP). Risk factors seem to explain some, but not all, of the observed relationships in FHS.

```{r rf-adjust}
run_fhs_adjusted <- function(cpgSet) {
  ewasModel_fhs_rfAdjust <- paste0("survObj~meth+sex+age+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+",
                                   "ldl+hdl+tg+sbp+hscrp+center+",
                                   paste0("cpPC",1:7,collapse="+"))
  fhsResAdjList <- foreach(meth=iter(betas_fhs[cpgSet,], by="row"), .packages="survival") %do%
    run_coxModel(meth, nonMethData[nonMethData$study=="fhs",], ewasModel_fhs_rfAdjust)
  fhsResAdj <- do.call(rbind, fhsResAdjList) %>%
    data.frame(stringsAsFactors=F, check.names=F) %>%
    dplyr::rename(p=`Pr(>|z|)`) %>%
    mutate_at(c("coef","z","p"), as.numeric)
  fhsResAdj
}
```
```{r combp-plots, warning=F, fig.cap="DMRs identified by Comb-p in WHI and validated in FHS at the (a) APOB and (b) SLC9A1 loci. Negative logarithms of EWAS p-values are shown as a function of the genomic coordinate. EWAS p-values from WHI are in red, FHS in green, and FHS adjusted for traditional CVD risk factors (total cholesterol, HDL-cholesterol, triglycerides, and systolic blood pressure) in blue. Dotted lines demarcate the DMR boundaries."}
make_bed <- function(resDF) {
  annoResDF <- resDF %>%
    inner_join(anno450k, by=c("CpG"="Name")) %>%
    mutate(start=as.numeric(pos), 
           end=as.numeric(pos),
           negLogP=-log10(p)) %>%
    dplyr::select(chr, start, end, negLogP, CpG)
}

resAnnoWhi <- make_bed(whiResFull)
resAnnoFhs <- make_bed(fhsResFull)

## ApoB locus
dmr <- c(combpResTbl$start[1], combpResTbl$end[1])
chrom <- combpResTbl$chr[1]
chromstart <- dmr[1]-2000
chromend <- dmr[2]+5000
resAnnoRegionWhi <- filter(resAnnoWhi, chr==chrom, start>=chromstart, end<=chromend)
resAnnoRegionFhs <- filter(resAnnoFhs, chr==chrom, start>=chromstart, end<=chromend)
resAnnoRegionFhsAdj <- mutate(resAnnoRegionFhs, negLogP=-log10(run_fhs_adjusted(CpG)$p))
plotDF <- bind_rows(list(resAnnoRegionWhi, resAnnoRegionFhs, resAnnoRegionFhsAdj), .id="Dataset") %>%
  mutate(Dataset=case_when(Dataset==1~"WHI", Dataset==2~"FHS", Dataset==3~"FHS (RF-adj)"),
         Dataset=factor(Dataset, levels=unique(Dataset)))
apobPlt <- ggplot(plotDF, aes(x=start/1000, y=negLogP, color=Dataset)) + 
  geom_point() +
  geom_vline(xintercept=dmr[1]/1000, linetype="dashed") +
  geom_vline(xintercept=dmr[2]/1000, linetype="dashed") +
  xlab("Position on chr. 2 (kb)") +
  ggtitle("ApoB locus")

## SLC9A1 locus
dmr <- c(combpResTbl$start[3], combpResTbl$end[3])
chrom <- combpResTbl$chr[3]
chromstart <- dmr[1]-5000
chromend <- dmr[2]+5000
resAnnoRegionWhi <- filter(resAnnoWhi, chr==chrom, start>=chromstart, end<=chromend)
resAnnoRegionFhs <- filter(resAnnoFhs, chr==chrom, start>=chromstart, end<=chromend)
resAnnoRegionFhsAdj <- mutate(resAnnoRegionFhs, negLogP=-log10(run_fhs_adjusted(CpG)$p))
plotDF <- bind_rows(list(resAnnoRegionWhi, resAnnoRegionFhs, resAnnoRegionFhsAdj), .id="Dataset") %>%
  mutate(Dataset=case_when(Dataset==1~"WHI", Dataset==2~"FHS", Dataset==3~"FHS (RF-adj)"),
         Dataset=factor(Dataset, levels=unique(Dataset)))
slc9a1Plt <- ggplot(plotDF, aes(x=start/1000, y=negLogP, color=Dataset)) + 
  geom_point(size=2) +
  geom_vline(xintercept=dmr[1]/1000, linetype="dashed") +
  geom_vline(xintercept=dmr[2]/1000, linetype="dashed") +
  xlab("Position on chr. 1 (kb)") +
  ggtitle("SLC9A1 locus")

plot_grid(apobPlt, slc9a1Plt, nrow=2, labels="auto")
```

## Weighted correlation network approach 

```{r wgcna, cache=1, warning=F}
load("../int/wgcnaRes.RData")

## Modules associated with incident CVD
egModelRes.list <- lapply(eigengenes, function(eg) {
  cox.fit <- coxph(Surv(time, event)~eg+age+race+bmi+smk_now+smk_py+CD4T+CD8T+Bcell+NK+Mono+Gran+dnaPull, 
                   data=nmd_whi)
  summary(cox.fit)$coef["eg",c("coef","z","Pr(>|z|)")]
})
egModelRes.df <- do.call(rbind, egModelRes.list) %>%
  as.data.frame() %>%
  rownames_to_column(var="module") %>%
  arrange(`Pr(>|z|)`)
sigModules <- gsub("^ME", "", egModelRes.df$module[egModelRes.df$`Pr(>|z|)`<0.05])

Zsummary.pres <- mp$preservation$Z$ref.whi$inColumnsAlsoPresentIn.fhs[sigModules,"Zsummary.pres"]
Zsummary.qual <- mp$quality$Z$ref.whi$inColumnsAlsoPresentIn.fhs[sigModules,"Zsummary.qual"]

tssNoIslands <- anno450k %>% 
  dplyr::rename(cpg=Name, group=UCSC_RefGene_Group) %>%
  dplyr::select(cpg, group, Relation_to_Island) %>%
  separate_rows(group, sep=";") %>% 
  filter(group %in% c("TSS200","TSS1500"), 
         Relation_to_Island!="Island")
wgcnaInputCpgs <- rownames(betas_whi)[rownames(betas_whi) %in% tssNoIslands$cpg]

fhsEigengenes <- lapply(setNames(sigModules,sigModules), function(col) {
  pc1rots <- prcomp(t(betas_whi[wgcnaInputCpgs[whiNet$colors==col],]), scale.=T)$rot[,"PC1"]
  commonCpgs <- intersect(rownames(betas_fhs), names(pc1rots))
  as.vector(t(betas_fhs[commonCpgs,]) %*% pc1rots[commonCpgs])
})
egModelResFhs.list <- lapply(fhsEigengenes, function(eg) {
  cox.fit <- coxph(Surv(time, event)~eg+age+bmi+sex+CD4T+CD8T+Bcell+NK+Mono+Gran+center+cpPC1+cpPC2+cpPC3+cpPC4+cpPC5+cpPC6+cpPC7, 
                   data=nmd_fhs)
  summary(cox.fit)$coef["eg",c("coef","z","Pr(>|z|)")]
})
egModelResFhs.df <- do.call(rbind, egModelResFhs.list) %>%
  as.data.frame() %>%
  rownames_to_column(var="module")

# mp3cpgAnnot <- anno450k %>%
#   dplyr::rename(CpG=Name, Gene=UCSC_RefGene_Name) %>%
#   filter(CpG %in% mp3cpgs) %>%
#   select(CpG, Gene, chr, pos)
# kable(mp3cpgAnnot)

# mp3 -> no association w/ only technical
# pink -> age abolishes
# blue -> no association w/ only technical
# lightyellow -> age abolishes

moduleEnrichment <- lapply(setNames(sigModules,sigModules), function(m) {
  gseaRes <- missMethyl::gometh(as.vector(wgcnaInputCpgs[which(whiNet$colors==m)]), all.cpg=wgcnaInputCpgs)
  data.frame(gseaRes) %>%
    arrange(P.DE) %>%
    head(5) %>%
    mutate(module=m) %>%
    select(module, Term, Ont, N, DE, P.DE, FDR)
})
allModuleEnrichment <- do.call(rbind, moduleEnrichment)

# mp3GsaTbl <- gometh(as.vector(mp3cpgs), collection="GO")
# kable(arrange(filter(mp3GsaTbl, FDR<0.1), FDR))
```

Next, we used weighted gene network correlation analysis (WGCNA) to further assess the association of groups of CpG sites with incident CVD events without requiring their regional proximity. WGCNA uses weighted correlation networks to identify highly correlated submodules that may be more interpretable and more robust as a group than the individual components. In order to focus on sites that are well-annotated, variable, and have better understood directional effects on transcription, only CpGs within 1.5kb of gene transcription start sites (n=`r length(wgcnaInputCpgs)`) were included.

The WGCNA analysis in WHI returned `r ncol(eigengenes)` modules, `r length(sigModules)` of which were at least nominally associated with incident events after adjustment for the same set of covariates as in the initial EWAS. Based on gene enrichment analyses, these modules associated with extracellular matrix function (pink), developmental processes (blue), and cholesterol binding and metabolism (lightyellow), while the fourth (mediumpurple3) consisted of CpGs in a single locus on chromosome 6 near RNF5P1 and AGPAT1. Based on thresholds suggested by the authors, z-scores calculated for these four modules in FHS showed strong preservation (all z-scores >10). However, two of these modules (mediumpurple3 and blue) did not show an association with incident CVD events after adjustment for technical covariates, and the other two (pink and lightyellow) did not show association after additional adjustment for age.

## Incorporation of genetic and epigenetic annotations improves mechanistic interpretation

```{r epigenomic-anotation, cache=1}
anno450kRanges <- GRanges(anno450k$chr, IRanges(start=anno450k$pos, end=anno450k$pos))

read_epiAnno <- function(broadPeaksFile) {
  # Given a path to a broadPeaks annotation file, read in the set of regions and output a GRanges object
  epiAnno.df <- suppressMessages(read_tsv(broadPeaksFile, col_names=F, progress=F)) %>%
    select(1, 2, 3, ncol(.)) %>%
    setNames(c("chr","start","end","negLogQ")) %>%
    filter(negLogQ>2)
  GRanges(seqnames=epiAnno.df$chr, IRanges(start=epiAnno.df$start, end=epiAnno.df$end))
}

test_epiAnno <- function(epiAnno.gr, ewasRes) {
  # Given a set of peak calls/regions of interest for an epigenomic annotation of interest (as GRanges 
  # object), return the -log(p) testing whether a set of EWAS results are associated with the annotation
  cpgsInAnno <- findOverlaps(anno450kRanges, epiAnno.gr)
  annoWithMembership <- mutate(anno450k, member=1:nrow(anno450k) %in% queryHits(cpgsInAnno))
  annoWithMembershipAndPvals <- inner_join(ewasRes, annoWithMembership, by=c("CpG"="Name"))
  tryCatch(unname(t.test(negLogP~member, data=annoWithMembershipAndPvals)$statistic),
           error=function(e) NA)
}

test_epiAnno <- function(epiAnno.gr, ewasRes) {
  # Given a set of peak calls/regions of interest for an epigenomic annotation of interest (as GRanges 
  # object), return the -log(p) testing whether a set of EWAS results are associated with the annotation
  cpgsInAnno <- findOverlaps(anno450kRanges, epiAnno.gr)
  annoWithMembership <- mutate(anno450k, member=1:nrow(anno450k) %in% queryHits(cpgsInAnno))
  annoWithMembershipAndPvals <- inner_join(ewasRes, annoWithMembership, by=c("CpG"="Name"))
  tryCatch({
    p <- phyper(sum(annoWithMembershipAndPvals$member*(annoWithMembershipAndPvals$negLogP>-log10(0.05))),
                sum(annoWithMembershipAndPvals$member),  # Number of sites in peaks
                sum(!annoWithMembershipAndPvals$member),  # Number of sites not in peaks
                sum(annoWithMembershipAndPvals$negLogP>-log10(0.05)))  # Number of sites with p<0.05
    -log10(min(p, 1-p))
  },
  error=function(e) NA)
}

assemble_epiAnno <- function(annoType, cellType, ewasRes) {
  # Given an epigenomic annotation type (e.g. H3K4me1), output a named list of GRanges objects
  annoDir <- paste0("../data/literature/roadmap/", annoType)
  annoFile <- grep(cellType, list.files(annoDir, full.names=T), value=T)
  if (length(annoFile)==0) return(NA) else epiAnno.gr <- read_epiAnno(annoFile)
  tStat <- test_epiAnno(epiAnno.gr, ewasRes)
  data.frame(EpiFeature=annoType, CellTypeCode=cellType, tStat=as.numeric(tStat))
}

epiFeatures <- list.files("../data/literature/roadmap/", pattern="^[DH].*")
cellTypeCodes <- unique(gsub("^.*/|-.*", "", list.files("../data/literature/roadmap/", recursive=T)))
featureCellGrid <- expand.grid(EpiFeature=epiFeatures, CellTypeCode=cellTypeCodes)
mergedEwasRes <- inner_join(whiResFull, fhsResFull, by="CpG", suffix=c(".whi",".fhs")) %>%
  mutate(negLogP=-(log10(p.whi)+log10(p.fhs))/2) %>%
  select(CpG, negLogP)

cl <- makePSOCKcluster(detectCores())
registerDoParallel(cl)
tStats <- foreach(anno=iter(featureCellGrid, by="row"), .combine=rbind,
                      .packages=c("tidyverse","GenomicRanges")) %dopar%
  assemble_epiAnno(anno$EpiFeature, anno$CellType, mergedEwasRes) 
stopCluster(cl)
```
```{r epigenomic-annotation-plots, fig.cap="Inference of cell type-specificity of EWAS signal. The heatmap shows t-statistics from t-tests comparing EWAS -log10(p) values based on membership in peaks for a given combination of cell type and epigenomic feature. The barplot shows the marginal distribution of t-statistic magnitudes (sum of absolute values) across all epigenomic features for each cell type."}
roadmapMeta <- read_csv("../data/literature/roadmap/jul2013.roadmapData.qc\ -\ Consolidated_EpigenomeIDs_QC.csv", col_types=cols())

epiAnnoPltData <- tStats %>%
  na.omit() %>%
  mutate(CellType=roadmapMeta$`Standardised epigenome name`[match(CellTypeCode,roadmapMeta$EID)]) %>%
  group_by(CellType) %>%
  mutate(cellTypeAbsMean=mean(abs(tStat), na.rm=T)) %>%
  filter(all(epiFeatures %in% EpiFeature)) %>%
  ungroup() %>%
  arrange(desc(cellTypeAbsMean)) %>%
  mutate(CellType=factor(CellType, levels=unique(CellType)))
heat <- ggplot(epiAnnoPltData, aes(x=EpiFeature, y=forcats::fct_rev(CellType), fill=tStat)) +
  geom_tile() +
  scale_fill_gradient2(low="blue", mid="white", high="red") +
  theme(axis.title=element_blank(), axis.text.x=element_text(size=9, angle=25, hjust=0.75),
        axis.text.y=element_text(size=9), axis.ticks.y=element_blank(), axis.line.y=element_blank(),
        legend.title=element_blank(), legend.text=element_text(size=9), legend.position="top")

epiAnnoSummary <- tStats %>%
  na.omit() %>%
  mutate(CellType=roadmapMeta$`Standardised epigenome name`[match(CellTypeCode,roadmapMeta$EID)]) %>%
  group_by(CellType) %>%
  filter(all(epiFeatures %in% EpiFeature)) %>%
  summarise(CellTypeAbsMean=sum(abs(tStat))) %>%
  arrange(desc(CellTypeAbsMean)) %>%
  mutate(CellType=factor(CellType, levels=unique(CellType)))

hist <- ggplot(epiAnnoSummary, aes(x=forcats::fct_rev(CellType), y=CellTypeAbsMean)) +
  geom_bar(stat="identity") +
  labs(y="Average t-statistic magnitude") +
  theme(axis.title=element_blank(), axis.text=element_blank(), 
        axis.ticks=element_blank(), axis.line=element_blank(),
        panel.background=element_blank()) +
  coord_flip()

# https://statbandit.wordpress.com/2011/07/29/a-ggplot-trick-to-plot-different-plot-types-in-facets/
  
plot_grid(heat, hist, align="h", axis="bt", ncol=2, rel_widths=c(4,1))
```

To probe the cell type specificity of the observed associations, a novel procedure based on epigenomic annotations was used. Peak calls for a series of epigenomic features (DNase hypersensitivity sites and multiple histone marks) have been determined as previously described [@Kundaje2014] and are available from the Roadmap Epigenomics database. These data were retrieved for a broad array of cell types in order to assign CpGs analyzed here to cell- and feature-specific peaks. For each annotation, t-tests were performed to compare the significance of sites ($-log_{10}(p)$ from EWAS) to their membership in a peak. Overall, this method would be expected to show greater significance in cell types relevant to the observed signal, whether as a specific component of a mixed tissue (e.g. blood) or as a surrogate for the relevant tissue.

In this cross-cohort analysis, monocyte annotations showed the strongest associations with EWAS signal across epigenomic features, with other blood cell types showing many of the other top ranks. Non-blood cell types also showing high ranks, potentially representative of blood acting as a surrogate tissue, included fibroblast, mesenchymal stem cells, and fetal tissues.

```{r gwas-integration, include=F}
cardiogram <- read_tsv("../data/literature/CARDIoGRAMplusC4D_1k.txt") 
cardiogramSig <- cardiogram %>%
  dplyr::rename(snp=markername, pos=bp_hg19, p=p_dgc) %>%
  filter(p<1e-8) %>%
  mutate(chr=paste0("chr",chr)) %>%
  select(snp, chr, pos, p)
cSigRanges <- GRanges(cardiogramSig$chr, IRanges(start=cardiogramSig$pos-1e6, end=cardiogramSig$pos+1e6))

suggestiveAnno <- select(anno450k[match(bothSuggestiveSites, anno450k$Name),], Name, chr, pos)
suggestiveRanges <- GRanges(suggestiveAnno$chr, IRanges(start=suggestiveAnno$pos, end=suggestiveAnno$pos))
siteOverlaps <- findOverlaps(suggestiveRanges, cSigRanges)

dmrAnno <- combpResTbl
dmrRanges <- GRanges(dmrAnno$chr, IRanges(start=dmrAnno$start, end=dmrAnno$end))
dmrOverlaps <- findOverlaps(dmrRanges, cSigRanges)
```

The positions of the `r length(bothSuggestiveSites)` DMPs and 4 DMRs were assessed for their proximity the set of genome-wide significant (p<1e-8) variants from the CaRDIoGRAMplusC4D meta-analysis [@Nikpay2015] (here, within 1MB of a GWAS hit). One DMP, `r suggestiveAnno$Name[queryHits(siteOverlaps)[1]]` in the gene body of PPP1R13L, was found to be about 500kb downstream of a series of hits in the TOMM40-APOE-C1-C2-C4 cluster on chromosome 19. Hi-C chromatin interaction data visualized using the 3DIV database [@Yang2017] indicated that this DMP and SNP cluster mark the approximate respective boundaries of a topologically associating domain (TAD) in the GM12787 lymphoblastoid cell line. In a some of the top-ranking cell types from the cell type-specific analysis above, this same TAD did not exist, although specific long-range chromatin contacts between these SNPs and the DMP were found (for example, in lung fibroblasts and mesenchymal stem cells).

```{r mr-prep, cache=1, include=F}
lead_cpg_from_region <- function(chr, start, end) {
  # Given a region, find all 450k CpGs in the region and return the top EWAS hit from that set
  regionCpgs <- filter(anno450k, chr==chr, pos>=as.numeric(start), pos<=as.numeric(end))$Name
  arrange(filter(whiResFull, CpG %in% regionCpgs), p)$CpG[1]
}
dmrLeadCpgs <- apply(combpResTbl, 1, function(row) 
  lead_cpg_from_region(row["chr"], row["start"], row["end"]))
names(dmrLeadCpgs) <- paste0(dmrLeadCpgs,"-",combpResTbl$annot_gene,"dmr")
names(dmrLeadCpgs)[4] <- paste0(dmrLeadCpgs[4],"-","Chr5dmr")

find_genes_byCpg <- function(cpg.vec) {
  map_chr(cpg.vec, function(cpg) {
    cpgAnnot <- filter(anno450k, Name==cpg)
    gsub(";.*", "", cpgAnnot$UCSC_RefGene_Name[1])
  })
}
names(bothSuggestiveSites) <- paste(bothSuggestiveSites, find_genes_byCpg(bothSuggestiveSites), sep="-")

topCpgs <- c(bothSuggestiveSites, dmrLeadCpgs)
# names(leadCpgs) <- paste(leadCpgs, find_genes_byCpg(leadCpgs), sep="-")
#   setNames(c(bothSuggestiveSites,dmrLeadCpgs),
#                     c(bothSuggestiveSites,paste0("dmr",combpResTbl$annot_gene)))

prep_mr <- function(allCpgs) {
  ewasRes <- whiResFull %>%
    dplyr::rename(cpg=CpG, ewasP=p) %>%
    mutate(ewasSign=sign(coef)) %>%
    select(cpg, ewasSign, ewasP) %>%
    filter(cpg %in% allCpgs)  # Reduce to only CpGs in set of interest to reduce size
  ewasCpGs <- anno450k %>%
    dplyr::rename(cpg=Name, cpgPos=pos) %>%
    mutate(cpgChr=gsub("chr", "", chr)) %>%
    select(cpg, cpgChr, cpgPos)
  mqtlSummary <- read_tsv("../data/literature/mqtldb.tab") %>%
    dplyr::rename(snp=SNP, cpg=gene, beta_mqtl=beta, fdr_mqtl=FDR) %>%
    mutate(se_mqtl=beta_mqtl/`t-stat`) %>%
    select(snp, cpg, beta_mqtl, se_mqtl, fdr_mqtl)
  mqtlSNPs <- read_tsv("../data/literature/ariesmqtlsnps.bim", col_names=F) %>%
    select(X2, X5) %>%
    setNames(c("snp","mqtlEffectAllele"))
  gwasSummary <- cardiogram %>%
    dplyr::rename(snp=markername, snpChr=chr, snpPos=bp_hg19, gwasEffectAllele=effect_allele,
                  beta_gwas=beta, se_gwas=se_dgc) %>%
    select(snp, snpChr, snpPos, gwasEffectAllele, beta_gwas, se_gwas)
  ewasRes %>%
    inner_join(ewasCpGs, by="cpg") %>%
    inner_join(mqtlSummary, by="cpg") %>%
    inner_join(mqtlSNPs, by="snp") %>%
    inner_join(gwasSummary, by="snp") %>%
    filter(cpgChr==snpChr) %>%  # Same chromosome for genetic IV
    mutate(beta_gwas=ifelse(mqtlEffectAllele==gwasEffectAllele, beta_gwas, -1*beta_gwas))  # Switch GWAS sign if effect alleles are reversed
}

mrPrep <- prep_mr(unique(unlist(topCpgs)))

snpSets <- lapply(topCpgs, function(lc) filter(mrPrep, cpg==lc)$snp)
noSnpSetNames <- names(snpSets[lengths(snpSets)==0])
oneSnpSetNames <- names(snpSets[lengths(snpSets)==1])
snpSetNames <- names(snpSets[lengths(snpSets)>1])

for (set in snpSetNames) write(snpSets[[set]], file=paste0("../int/", set, "_snps.txt"))
```
```{r mr-res, message=F}
mrSetsPruned <- lapply(setNames(snpSetNames, snpSetNames), function(set) 
  scan(paste0("../int/", set, "_snps_pruned.txt"), what=character()))

library(MendelianRandomization)

run_mr <- function(topCpg, snpSet, mrType="ivw") {
  instrumentData <- filter(mrPrep, cpg==topCpg & snp %in% snpSet)
  mrInput <- MendelianRandomization::mr_input(bx=instrumentData$beta_mqtl, bxse=instrumentData$se_mqtl,
                                              by=instrumentData$beta_gwas, byse=instrumentData$se_gwas,
                                              snps=instrumentData$snp)
  mr.plt <- mr_plot(mrInput, orientate=T, line=mrType, interactive=F)
  mr.fit <- if (mrType=="ivw") mr_ivw(mrInput) else mr_egger(mrInput)
  list(mrInput=mrInput, plt=mr.plt, fit=mr.fit)
}

mrResSingle <- lapply(oneSnpSetNames, function(set) run_mr(topCpgs[set], snpSets[[set]]))
mrRes <- lapply(setNames(snpSetNames, snpSetNames), 
                function(set) run_mr(topCpgs[set], mrSetsPruned[[set]]))
```
```{r apob-specific}
apobSiteName <- grep("APOB", snpSetNames, value=T)
apobEggerRes <- run_mr(topCpgs[apobSiteName], mrSetsPruned[[apobSiteName]], mrType="egger")

cardiogramSig <- cardiogram %>%
  dplyr::rename(snp=markername, pos=bp_hg19, p=p_dgc) %>%
  filter(p<1e-5) %>%
  mutate(chr=paste0("chr",chr)) %>%
  select(snp, chr, pos, p)
cSigRanges <- GRanges(cardiogramSig$chr, IRanges(start=cardiogramSig$pos-1e6, end=cardiogramSig$pos+1e6))

apobDmrAnno <- combpResTbl[combpResTbl$annot_gene=="APOB",]
apobDmrRanges <- GRanges(apobDmrAnno$chr, IRanges(start=apobDmrAnno$start, end=apobDmrAnno$end))
apobDmrOverlaps <- findOverlaps(apobDmrRanges, cSigRanges)
apobSnpsBed <- cardiogramSig[subjectHits(apobDmrOverlaps),] %>%
  filter(!is.na(as.numeric(pos))) %>%
  mutate(start=pos, end=pos, negLogP=-log10(p)) %>%
  select(chr, start, end, snp, negLogP)
cat('track name="CVD SNPs near APOB" description="CARDIoGRAMplusC4D suggestive SNPs within 1MB of APOB DMR"', file="../int/apob_proximal_cvd_snps.bed")
write_tsv(apobSnpsBed, "../int/apob_proximal_cvd_snps.bed", col_names=F, append=T)
```

Though the prospective nature of this EWAS analysis precludes reverse causation, assocations between DMPs/DMRs and CVD are still subject to confounding. To better examine the causality of these relationships, a Mendelian randomization (MR) approach using single nucleotide polymorphisms (SNPs) as instrumental variables was undertaken through the use of external datasets. SNP-methylation associations were taken from the mQTLdb (ARIES study) [@Gaunt2016] and SNP-CVD associations were taken from the CARDIoGRAMplusC4D meta-analysis as above. 

The inverse-variance weighted method was used to assess the causal relationship between differential methylation and potential pleiotropy of the instrumental variables (here, SNPs) for the `r length(bothSuggestiveSites)` DMPs and 4 DMRs. The most significant CpG site from each region was chosen as the "lead CpG" to be used in the MR analysis. `r length(noSnpSetNames)` methylation loci lacked links to any SNPs in the mQTLdb. Of the remaining set of CpGs, one (`r topCpgs[grepl("APOB",names(topCpgs))]`, from the APOB region) demonstrated a consistent and "dose-dependent" association of its IV SNPs with CVD (p=`r as.character(signif(mrRes[[grep("APOB", snpSetNames, value=T)]]$fit$Pvalue,2))`; Fig. \@ref(fig:genomic-annotation-figure)a). MR-Egger regression [@Bowden2015] was then applied to address potential pleiotropy of the IV SNPs, and showed no direct evidence of pleiotropy (MR-Egger intercept p=`r as.character(signif(apobEggerRes$fit$Pvalue.Int,2))`), though the causal relationship estimate was attenuated (MR-Egger slope p=`r as.character(signif(apobEggerRes$fit$Pvalue.Est,2))`).

Based on this result, we expanded the previous CVD GWAS-related analysis to include genome-wide "suggestive" SNPs (p<1e-5) and found that the APOB DMR was within 1MB of `r length(subjectHits(apobDmrOverlaps))` of these (non-independent) loci. Because this gene is expressed primarily in liver and small intestine, we examined Hi-C interaction data as above for this locus in these two tissues as well as the GM12787 lymphocyte precursor. This region was contained by a TAD in the two tissues of interest (liver and small bowel) but not in the blood cell precursor (Fig. \@ref(fig:genomic-annotation-figure)b).

```{r genomic-annotation-figure, fig.asp=1.2, fig.cap="Evidence for a causal association of APOB methylation with CVD. (a) Plot of the strength of genetic associations with CVD vs. with APOB methylation for the series of instrumental variables used in the Mendelian randomization analysis. The trend line shown is the Egger regression fit. (b) Genomic annotations for the area containing the APOB gene and all suggestive GWAS loci within 1MB. Heatmap intensities represent interaction frequencies in small bowel."}
# include a figure showing anything from the GWAS linkage along with the egger regression plot
# motivate using GWAS then confirm causality using regression? --> single figure
apobEggerPlt <- mr_plot(apobEggerRes$mrInput, line="egger", orientate=T, interactive=F) +
  theme(axis.title=element_text(size=10))
# apobRegionImg <- png::readPNG("../int/apob_region_smallBowel.png")
apobRegionImg <- png::readPNG("../int/apob_region_viz.png")
apobRegionGrob <- grid::rasterGrob(apobRegionImg, interpolate=T)
plot_grid(apobEggerPlt, apobRegionGrob, nrow=2, rel_heights=c(3,2), labels="auto")
# grid.arrange(apobEggerPlt, apobRegionGrob, heights=c(3,5), nrow=2)
```

(Note to readers: This regional plot is just a stand-in -- recommendations for publication-quality/nice-looking browsers to incorporate annotations, TADs, etc. are welcome.)

To further probe the mechanistic relationship between methylation levels and CVD-related biological activity, the SEPIRA method [@Chen2017] was used to infer blood-specific transcription factor activity. First, blood-specific transcription factors are determined using expression data (here, from GTEx), then levels of these TFs in each sample are inferred from DNA methylation data.

```{r sepira, warning=F, out.width="70%", fig.cap="Relationships between estimated blood-specific transcription factor activities and incident CVD in both datasets. Shown are coefficients from regressing inferred TF levels on binary incident event along with all EWAS covariates. Only those TFs with nominal p<0.05 for at least one cohort are shown."}
load("../int/sepira_results.RData")  # Contains actTF_whi, actTF_fhs, and net.o

whiTFres <- apply(actTF_whi, 1, function(tfAct)
  summary(lm(tfAct~event+age+race+smk_now+smk_py+CD4T+CD8T+NK+Bcell+Mono+Gran+dnaPull, 
             data=nmd_whi))$coef["eventTRUE",c("Estimate","Pr(>|t|)")])
fhsTFres <- apply(actTF_fhs, 1, function(tfAct)
  summary(lm(tfAct~event+age+sex+smk_now+smk_py+CD4T+CD8T+NK+Bcell+Mono+Gran+center+cpPC1+cpPC2+cpPC3+cpPC4+cpPC5+cpPC6+cpPC7, 
             data=nmd_fhs))$coef["eventTRUE",c("Estimate","Pr(>|t|)")])

whiTfDf <- tibble(dataset="WHI", gene=colnames(whiTFres), coef=whiTFres[1,], p=whiTFres[2,]) %>%
  mutate(fdr=p.adjust(p, method="BH"))
fhsTfDf <- tibble(dataset="FHS", gene=colnames(fhsTFres), coef=fhsTFres[1,], p=fhsTFres[2,]) %>%
  mutate(fdr=p.adjust(p, method="BH"))
plotDF <- bind_rows(whiTfDf, fhsTfDf) %>%
  filter(gene %in% unique(gene[p<0.05]))

ggplot(plotDF, aes(x=gene, y=as.numeric(coef), fill=dataset)) + 
  geom_bar(stat="identity", position="dodge", width=0.5) + 
  ylab("Regression coefficient") +
  # ggtitle("Estimated TF activities regressed on incident event (binary) + covariates") +
  theme(axis.text.x=element_text(angle=40), axis.title.x=element_blank())
```

A set of `r ncol(whiTFres)` blood-specific transcription factors were found based on the input expression data from GTEx. Activity profiles for this set of TFs were then inferred from DNA methylation data for each sample. No TFs reached significance and replication across datasets, but two (TBX21 & MTA2) showed moderate effects and consistency in direction. 

## Comparison to known risk factors

```{r rf-comparison}
# dmrEigencpgs <- apply(combpResTbl, 1, function(row) {
#   cpgs <- anno450k$Name[anno450k$chr==row$chr, anno450k$pos>=row$start, anno450k$pos<=row$end]
#   prcomp(t(betas_fhs[cpgs,]), scale.=T)$x[,"PC1"]
# })
# sigFeatures <- cbind(betas_fhs[bothSuggestive,], do.call(cbind, c(dmrEigencpgs)))

topCpgBetasAdj <- t(map_dfc(setNames(topCpgs,topCpgs), function(cpg) {
  lm(betas_fhs[cpg,] ~ center+cpPC1+cpPC2+cpPC3+cpPC4+cpPC5+cpPC6+cpPC7, data=nmd_fhs)$residuals
}))

sigByRF <- expand.grid(cpg=topCpgs, rf=c("ldl","hdl","tg","sbp","hscrp","diabetes"))
sigByRF$cor <- apply(sigByRF, 1, function(row) {
  cor.test(topCpgBetasAdj[row["cpg"],], as.numeric(nmd_fhs[[row["rf"]]]))$estimate
})
sigByRF$cpgLabel=factor(sigByRF$cpg, levels=topCpgs, labels=names(topCpgs))

rfPlt <- ggplot(sigByRF, aes(x=rf, y=cpgLabel, fill=cor)) +
  geom_tile(color="white") +
  scale_fill_gradient2(low='#0000FF',mid='#FFFFFF',high='#FF0000') +
  # ggtitle("Correlations of DMPs with traditional risk factors") +
  theme(legend.title=element_blank(), axis.title=element_blank())
```

```{r partial-correlation-network, message=F, include=F}
library(qgraph)

# sbp_cpgs <- c(PHGDH="cg14476101", TXNIP="cg19693031")
# hdl_cpgs <- c(DHCR24="cg17901584")#, ABCG1="cg06500161")
# tg_cpgs <- c(CPT1A_1="cg00574958", CPT1A_2="cg17058475")
# tc_cpgs <- c(SQLE="cg00285394", NLRC5="cg07839457")

sbp_cpgs <- c(PHGDH="cg14476101")
hdl_cpgs <- c(DHCR24="cg17901584")
tg_cpgs <- c(CPT1A_1="cg00574958")
tc_cpgs <- c(SQLE="cg00285394")

acs_cpgs <- c(RUNX3="cg21217270", DGKQ="cg12226453", RGS3="cg14049634", ANKRD11="cg27513667")
cvdSR_cpgs <- c(GRIP1="cg09414535", KCNJ14="cg10222534", PKD2="cg26215428", AIM2="cg10636246",
                HRH2="cg14345676", NGEF="cg19485804", TNS1="cg18752854")

allRFs <- c("age","sex","bmi","diabetes","chol","ldl","hdl","tg","sbp","glu","hscrp","smk_now")
allCpgs <- c(sbp_cpgs, hdl_cpgs, tg_cpgs, tc_cpgs, acs_cpgs, cvdSR_cpgs, topCpgs)

includeRFs <- c("diabetes","ldl","hdl","tg","sbp","hscrp")
includeLitCpgs <- c()
includeEwasCpgs <- topCpgs
names(includeEwasCpgs) <- gsub(".*-","",names(includeEwasCpgs))

corInputData <- cbind(select(nmd_fhs, one_of(allRFs)), t(topCpgBetasAdj)) %>%
  na.omit() %>%
  mutate(sex=sex=="M") %>%
  mutate_all(as.numeric) %>%
  select(c(includeRFs, includeLitCpgs, includeEwasCpgs)) %>%
  setNames(c(includeRFs, names(includeLitCpgs), natmes(includeEwasCpgs)))

corMat <- cor(corInputData)
# sparsePcMat <- EBICglasso(corMat, n=nrow(corInputData), lambda.min.ratio=0.0001)
# qgraph(sparsePcMat, layout="spring", 
#        groups=rep(c("rf","litCpg","ewasCpg"),
#                   times=c(length(includeRFs),length(includeLitCpgs),length(includeEwasCpgs))),
#        label.color="white", label.cex=1.3, labels=names(corInputData))
png("../output/figures/pcPlt.png")
qgraph(corMat,
       graph="pcor", threshold="bonferroni", sampleSize=nrow(corInputData),
       groups=rep(c("Risk Factor","litCpg","EWAS CpG"),
                  times=c(length(includeRFs),length(includeLitCpgs),length(includeEwasCpgs))),
       layout="spring",
       labels=names(corInputData), label.color="white", label.cex=1.25)
# pcPlt <- recordPlot()
dev.off()
```

```{r rf-cor-plots, fig.asp=1.5, fig.cap="Relationships between methylation and traditional CVD risk factors. (a) Pearson correlations between DMPs and DMR lead CpGs and risk factors. (b) Partial correlation network between and among the same CpGs and risk factors. Edge widths correspond to partial correlation magnitudes, and edge colors to partial correlation sign (green positive; red negative). CpG sites are labeled based on their associated genes, and CpGs representing DMRs are labeled as such."}
# grid.arrange(rfPlt, pcPlt, nrow=2)
pcPlt <- grid::rasterGrob(png::readPNG("../output/figures/pcPlt.png"), interpolate=T)
plot_grid(rfPlt, pcPlt, nrow=2, rel_heights=c(1,2), labels="auto")
# https://stackoverflow.com/questions/14124373/combine-base-and-ggplot-graphics-in-r-figure-window
# later answer doesn't require gridBase
```

To understand how these findings fit into the context of known CVD biology and risk assessment, we evaluated relationships between DMPs and DMRs (represented by their lead CpGs) with traditional blood-based CVD risk factors. First, simple correlations were examined between EWAS sites (single and lead CpGs from DMRs) and log-transformed risk factors (Fig. \@ref(fig:rf-cor-plots)a). A few sites, including cg06500161 (ABCG1 body) and cg02519286 (lead CpG from the SLC9A1 DMR) showed strong correlations with multiple risk factors. Others, such as cg11772801 (TMX1 body), showed little such correlation, indicating that these CpGs may relate to incident CVD through other mechanisms. 

Next, a partial correlation network was constructed to enable inspection of putative direct relationships between risk factors and the CpG sites uncovered here (Fig. \@ref(fig:rf-cor-plots)b). The network revealed a general clustering of known blood-based risk factors apart from CpG sites. Unlike the other risk factors, hsCRP demonstrated independent correlations with many CpG sites. Furthermore, a strongly associated cluster of CpGs annotated to genes MS4A1, APBA2, APLP2, and TMX1 was apparent. Triglyceride levels served as a central node in the network, linking to multiple CpG sites, as well as systolic blood pressure, lipid levels, and diabetes status.

```{r partial-correlation-network-noAge, message=F, eval=F}
library(qgraph)

# sbp_cpgs <- c(PHGDH="cg14476101", TXNIP="cg19693031")
# hdl_cpgs <- c(DHCR24="cg17901584")#, ABCG1="cg06500161")
# tg_cpgs <- c(CPT1A_1="cg00574958", CPT1A_2="cg17058475")
# tc_cpgs <- c(SQLE="cg00285394", NLRC5="cg07839457")

sbp_cpgs <- c(PHGDH="cg14476101")
hdl_cpgs <- c(DHCR24="cg17901584")
tg_cpgs <- c(CPT1A_1="cg00574958")
tc_cpgs <- c(SQLE="cg00285394")

acs_cpgs <- c(RUNX3="cg21217270", DGKQ="cg12226453", RGS3="cg14049634", ANKRD11="cg27513667")
cvdSR_cpgs <- c(GRIP1="cg09414535", KCNJ14="cg10222534", PKD2="cg26215428", AIM2="cg10636246",
                HRH2="cg14345676", NGEF="cg19485804", TNS1="cg18752854")

allRFs <- c("age","sex","bmi","diabetes","chol","ldl","hdl","tg","sbp","glu","hscrp")
allCpgs <- c(sbp_cpgs, hdl_cpgs, tg_cpgs, tc_cpgs, acs_cpgs, cvdSR_cpgs, topCpgs)

includeRFs <- c("diabetes","ldl","hdl","tg","sbp","hscrp")
includeLitCpgs <- c()
includeEwasCpgs <- dmrLeadCpgs
names(includeEwasCpgs) <- gsub(".*-","",names(includeEwasCpgs))

corInputData <- cbind(select(nmd_fhs, one_of(allRFs)), t(topCpgBetasAdj)) %>%
  na.omit() %>%
  mutate(sex=sex=="M") %>%
  mutate_all(as.numeric) %>%
  select(c(includeRFs, includeLitCpgs, includeEwasCpgs)) %>%
  setNames(c(includeRFs, names(includeLitCpgs), names(includeEwasCpgs)))

corMat <- cor(corInputData)
# sparsePcMat <- EBICglasso(corMat, n=nrow(corInputData), lambda.min.ratio=0.0001)
# qgraph(sparsePcMat, layout="spring", 
#        groups=rep(c("rf","litCpg","ewasCpg"),
#                   times=c(length(includeRFs),length(includeLitCpgs),length(includeEwasCpgs))),
#        label.color="white", label.cex=1.3, labels=names(corInputData))
pcPlt <- qgraph(corMat,
       graph="pcor", threshold="bonferroni", sampleSize=nrow(corInputData),
       groups=rep(c("rf","litCpg","ewasCpg"),
                  times=c(length(includeRFs),length(includeLitCpgs),length(includeEwasCpgs))),
       layout="spring",
       labels=names(corInputData), label.color="white", label.cex=1.25)
pcPlt
```

<!--
Next, a Bayesian network was created to link methylation at these CpG sites with both common risk factors and CpG sites previously established as related to these risk factors.  with common CVD RFs.....This approach has the advantage of providing a statistical inference of directionality to relationships between entities. We first undertook a preliminary analysis to validate the Bayesian network method, in which a structure was inferred between lipid-based risk factors and CpG sites previously established as related to these risk factors in EWAS. Not only did the expected links appear in the final graph structure, but they also strongly favored a directionality in which lipids affected methylation levels, as has been previously reported. Relevant CpGs from this study were then examined together with risk factors in a similar model. 
-->


# Discussion

Investigation of DNA methylation data in samples taken before onset of a disease addresses the central difficulty of reverse causation in epigenome analyses. Here, we looked for prospective biomarkers of CVD onset using survival models, and discovered a set of CpG sites and regions that are suggestively associated across the FHS and WHI datasets.

* basic CpG sites rundown -- have these been discovered elsewhere?

In order to better understand the effects of sex and race differences across the two datasets, we undertook stratified EWAS analyses. First, the top regression coefficients from WHI were compared to those from the same sites in men and women from FHS. Though women did show somewhat greater correlation of coefficients, neither stratum agreed as well as the full FHS dataset, indicating that sex is not a major driver of the observed heterogeneity in EWAS results. Next, an equivalent comparison was made between FHS coefficients and those from separate races in WHI. Here, coefficients from both whites and hispanic subjects showed greater agreement than the full population, while there was essentially no correlation in black subjects. Importantly, the group of hispanic subjects in WHI are mostly of Mexican descent [@Zambrana2014], and thus share a greater genetic similarity to those of European rather than African descent.
Previous observations have supported the generalizability of cardiovascular risk functions from the Framingham Heart Study to diverse populations [@Fox2016], and a reasonable overlap of genetic risk loci across ancestries [@Lettre2011]. However, the patterns observed here suggest caution in generalizing epigenetic results from FHS to other populations. Despite this intriguing pattern with respect to race, the improvement in correlation was modest, suggesting that demographic differences were not generally driving the observed differences in epigenetic associations.

* anything explicit about race dominating over sex in explaining epigenetic patterns?

Though no strong patterns were observed at the single-CpG level, we searched for differentially methylated regions, which may be more statistically robust and biologically meaningful [@Jaffe2012]. Four associations passed a strict multiple test correction threshold and were validated in FHS, the strongest of which was found in the ______ region of the APOB gene. There is existing support for both mechanistic [@Goff2014] and genetic [@Liu2015] associations between APOB and caridovascular disease, related to its function in cholesterol transport. Indeed, a recent Mendelian randomization-based analysis suggested a causal link between ____ the same ____ CpG and ApoB, the APOB protein product and known CVD risk factor [@Richardson2017]. APOB methylation has been linked to differential expression in liver cells [@ElTaghdouini2015], but whether this translates to increased circulating protein levels is not clear.

To expand on this observation, we undertook a Mendelian randomization analysis to examine potential causal relationships between the regions discovered here and cardiovascular disease, using existing databases for mQTLs (the same as was used in Richardson et al. for CVD risk factors) and CVD GWAS associations. We showed a possible causal relationship between methylation at the lead CpG from the APOB region and CVD. It is particularly interesting that this relationship was uncovered in whole blood samples, as the APOB gene is expressed primarily in small intestine and liver. Thus, in this case, blood may be acting as a surrogate reflecting methylation changes in the primary disease-relevant tissue.

Identification of specific cell types of interest may be valuable for understanding disease mechanisms as well as informing follow-up studies in specific cell types [@Bauer2018]. To further probe the cell-type specificity of our EWAS signal, we used a novel method based on the assumption that correlations between epigenomic feature annotations (e.g. histone marks) and EWAS signal strength will be stronger in relevant cell types. By averaging the strengths of this association across features for each cell type, we were able to rank cell types based on their potential relevance to our overall signal. As expected, blood cell types were all highly ranked. Monocytes showe the greatest importance based on this metric, supporting their well-known integral role in CVD pathogenesis [@Libby2006]. We also observed high ranks for fibroblasts and adrenal cells, both of which have been linked to CVD [@Murtha2017;@Rosmond2000], and which may suggest additional tissues for which blood is acting as a surrogate.

* WGCNA discussion

The methylation patterns observed here may be acting up- or downstream of known cardiovascular risk pathways (e.g. lipids), or may represent novel disease mechanisms. We looked at correlations between lead CpGs from our identified DMRs and classical CVD risk factors, including BMI, diabetes status, lipid levels, blood pressure, and hsCRP. None of the DMRs correlated strongly with these risk factors. Surprisingly, this included a lack of correlation between APOB methylation and cholesterol levels, which would otherwise be expected based on the biological function of the ApoB protein product. ....more when have bmi.... To move closer to mechanistic or causal relationships, we constructed a partial correlation network to display only Bonferroni-level correlations conditional on all other DMRs and risk factors. Most traditional risk factors showed strong connections to each other, to the exclusion of the DMRs. However, hsCRP stood out as a notable exception, showing stronger conditional correlations with methylation levels than with other risk factors. This suggests that these DMRs mediate or track disease pathways more related to inflammation than lipid levels, which may be intuitive given that they were assessed in blood samples.

There are a few important limitations to this association study. As discussed, only blood samples were analyzed, which represent an important tissue for CVD pathology but do not fully capture epigenetic dynamics of other disease-relevant tissues. Demographic heterogeneity of the two datasets may have contributed to a lack of sensitivity in identifying or validating population-specific epigenetic associations with CVD. Additionally, as this study examines relationships between baseline methylation levels and incident disease, the measurement of DNA methylation at only a single time point precluded an understanding of the stability of these epigenetic signatures over time.

* Summary para.

# Methods

## Study participants and phenotype collection

Data for the discovery set came from a combined case-control and pseudo case-cohort sampling of 2129 women from the Women's Health Initiative study. Included subjects had no self-reported CVD at baseline, and cases experienced an incident CHD event during follow-up. Data are available in the dbGaP public repository (accession: phs000200.v11.p3).

Data for the validation set came from a substudy of the Framingham Heart Study that measured DNA methylation in 2726 subjects from the Offspring Cohort. Fasting blood samples for both methylation and clinical biochemistry were collected from participants at Exam 8. Data are available in the dbGaP public repository (accession: phs000007.v29.p10).

More details on the cohorts will be requested from collaborators. 

CVD events were defined as adjudicated coronary heart disease events, stroke, or death from CVD.  

## DNA methylation data processing

In both cohorts, DNA methylation data were collected using the Illumina HumanMethylation450 microarray platform. (reference) Preprocessing was performed using the *minfi* and *wateRmelon* packages for R. As a quality control step, samples were removed if they showed weak overall signal based on visual inspection of an intensity plot, if they had more than 10% of probes undetected at a detection threshold of p<1e-16, or if the reported sex did not match the predicted sex based on methylation patterns. Probes were removed if they met any of the following criteria: more than 10% of samples undetected at a detection threshold of p<1e-16, location in the X or Y chromosomes, non-CpG probes, cross-hybridizing probes, probes measuring SNPs, and probes with an annotated SNP at the CpG site or in the single-base extension region. Samples were normalized using the Noob method for background correction and dye-bias normalization, followed by the BMIQ method for probe type correction. For each dataset, principal components analysis was performed on the set of control probes using code adapted from the CPACOR method of Lehne et al. [@Lehne2015] to account for technical variation. Blood cell counts for 6 blood cell types were estimated using the reference-based method of Houseman et al. [@Houseman2012]. 

## Association of DNA methylation with incident CVD events

After quality control and filtering steps, `r nrow(betas_whi)` (WHI) and `r nrow(betas_fhs)` (FHS) CpG sites
remained for downstream analysis. Visualization of the associations between top principal components for each dataset and major biological and technical variables was examined using heatmaps (Supp. Fig. \@ref(fig:supp-var-plots)). Based on these assessments as well as examination of Scree plots for CPACOR principal components, 

For each CpG site, a Cox proportional hazards regression was performed to assess the relationship between methylation beta value as the independent variable and time-to-event measures for incident CVD while adjusting for biological covariates (age, bmi, smoking status, smoking pack-years, and estimated blood cell counts; race in WHI; sex in FHS) and technical covariates (DNA pull batch in WHI; analysis center in FHS). To remove the influence of outliers in the regression, samples were removed from the analysis for each CpG if their beta value was outside of an extended outer fence (< 25%ile - $5*IQR$ or > 75%ile + $5*IQR$).
QQ plots revealed that genomic inflation was not adequately controlled in FHS, so 7 CPACOR principal components (chosen based on a Scree plot assessment) were additionally adjusted for in FHS.

For the main analysis of single sites, a specific subset of variable sites were selected based on a previous publication that performed bisulfite sequencing in isolated monocytes, CD4+ T-cells, and neutrophils [@Hachiya2017]. This study calculated "reference intervals", defined as the difference in beta value between the 5th and 95th percentile, as a measure of the variability of a given CpG site. Based on their provided recommendation, sites were included here if they demonstrated a reference interval of more than 30 in any cell type.

## Gene set enrichment analysis

Gene ontology-based enrichment analysis of DMPs was performed using the gometh function from the *missMethyl* package for R. In this procedure, gene annotations for CpG sites are retrieved, and enrichment analysis is performed while accounting for the bias in number of CpG probes per gene on the microarray. 

## DMR detection

Comb-p (implemented as a Python module) was used to call differentially methylated regions (DMRs) based on p-values from the EWAS, thus removing the requirement for additional covariate adjustment. Comb-p first calculates an autocorrelation function (ACF), for which a maximum distance of 1kb and a step size of 50 bases were used. Next, it uses the ACF to adjust each p-value using a Stouffer-Liptak-Kechris correction, followed by identification of contiguous regions of sites with adjusted p-values below some threshold (here, p<0.1 with no more than 500 bases between neighboring sites in a region). Finally, the ACF is recalculated out to the maximum region size (a step size of 50 was used here as well) and regional p-values are calculated using the Stouffer-Liptak test.

For multiple testing correction of DMRs, Comb-p calculates the number of effective tests as the number of loci tested divided by the size of the region (typically ~1kb). A more stringent correction was applied here, in which the number of independent tests was calculated as the number of loci tested divided by the number of probes in the region.

## Weighted gene correlation network analysis

Weighted gene correlation network analysis (WGCNA) was used to find highly correlated modules of CpG sites. The set of input CpGs, chosen to be generally the most interpretable and variable, were selected as those upstream of transcription start sites ("TSS200" and "TSS1500" based on Illumina's annotation file) but not directly in CpG islands ("Island" in the annotation file), resulting in a set of `r length(wgcnaInputCpgs)` sites. For computational tractability, blockwise module detection was performed, which treats block of features separately for network creation and module detection, followed by eventual merging of highly similar modules. For the main analysis performed in WHI, parameters included a block size of 20,000, a soft-thresholding power of 8, and a signed network type. Modules were assessed using the same regression framework and covariate set as was used in the initial EWAS, replacing single-site beta values with the first principal component of the module beta value matrix ("eigenCpG"). Enrichment analysis was carried out in each of the nominally associated modules using the method described above. Module preservation assessment in was completed in FHS based on z-scores, followed by calculation of eigenCpGs (according to the principal component weights from WHI) and assessment of associations with incident CVD. 

## Inference of transcription factor activity

The SEPIRA method [@Chen2017] was used to infer the activity of blood-specific transcription factors from DNA methylation data (code was adapted from Chen et al.). RNA-Seq data for multiple cell types was obtained from GTEx (release V7; TPM format) and used to infer a blood-specific regulatory network through comparison of transcription factor (TF) expression in blood cells vs. other cell types. Next, activity of these cell type-specific TFs was inferred after estimating cell type-specific gene activity from methylation as introduced in Jiao et al. [@Jiao2014] Briefly, methylation-based gene activities are estimated by assessing average methylation levels near transcription start sites of genes, with the assumption that methylation in these regions consistently associates with decreased activity. TF activity for each sample is then assessed by regressing these gene activity profiles on the TF gene target profile. These TF activities were then assessed in relation to incident CVD using the same Cox regression method and covariate adjustment as was used in the EWAS as well as for the modules from WGCNA above.

## Inference of cell type specificity

Epigenomic annotations were used to test for relative enrichment of strong EWAS signal in CpG sites located in peaks for epigenomic features in specific cell types. Annotations for broad peaks in DNase sensitivity as well as ChIP-seq signal for H3K4me1, H3K4me3, H3K27ac, H3K27me3, and H3K36me3 were obtained for `r nrow(epiAnnoSummary)` cell types from the NIH Roadmap Epigenomics Project database [[cite the specific 111 epigenomes paper]]. For each combination of epigenomic feature and cell type, CpGs from the HumanMethylation450 array were classified as to their membership in a peak locus. EWAS results from WHI and FHS were combined by taking the mean of their $-log_{10}(p)$ values. A t-test was then used to compare these significance values based on peak membership of the associated sites. T-statistics from these t-tests were used to assess the relative strength of the EWAS significance-peak membership associations in across cell types.

## Genetic data integration

Methylation-quantitative trait locus (mQTL) relationships were obtained from the mQTLdb (mqtldb.org). These relationships were generated by Gaunt et al. [@Gaunt2016] from an analysis of paired genotype and methylation data from the ARIES longitudinal study of mother-child pairs. Only relationships found at the "Middle Age" time point were used here. GWAS summary statistics for coronary artery disease and myocardial infarction (based on imputation using 1000 genomes phase 1 v3) were contributed by CARDIoGRAMplusC4D investigators [@Nikpay2015] and were downloaded from www.cardiogramplusc4d.org. 

Mendelian randomization analysis was performed using the R package *MendelianRandomization*. To generate the input for this analysis, SNPs (instrumental variables; IVs) were found in the mQTLdb that associated with the set of top DMPs and DMRs (exposures), with the most significant EWAS CpG from each DMR being used as the representative exposure variable. Only SNPs on the same chromosome as the CpG of interest were included. For CpGs with multiple associated mQTL SNPs, the list of SNPs was first pruned to correlations of r<0.2 using the NCI tool SNPclip (https://analysistools.nci.nih.gov/LDlink/?tab=snpclip). The inverse-variance weighted estimator (mr_ivw function), which assumes that all variants are valid instrumental variables, was used initially to calculate causal effect estimates. MR-Egger regression (mr_egger function) was then used to assess this instrumental variable assumption.

## Risk factor comparison

CVD risk factors were log-transformed to reduce the influence of extreme values on correlations. Methylation beta-values in FHS were adjusted for center and 7 control-probe principal components (as in the EWAS) using linear models, and the residuals were taken forward for use in the correlation analysis. Pearson correlations were computed between each risk factor and each CpG of interest (8 suggestive CpGs plus 4 DMR lead CpGs). Next, the full Pearson correlation matrix consisting of risk factors and these CpG sites was transformed into a partial correlation matrix using the R package *qgraph*. The partial correlation network was visualized using a network format after removing edges that did not pass a Bonferroni correction.

# Supplementary Data

\newcommand{\beginsupplement}{%
\setcounter{table}{0}
\renewcommand{\thetable}{S\arabic{table}}%
\setcounter{figure}{0}
\renewcommand{\thefigure}{S\arabic{figure}}%
}

\beginsupplement

```{r supp-var-plots, fig.asp=2, fig.cap="Significance of regression of top PCs from methylation beta value matrix on covariates."}
plot_grid(whiVarPlt, fhsVarPlt, nrow=2)
```

```{r supp-hachiya-plot, fig.cap="Reference interval comparison between variable sites as taken from Hachiya et al. and all other sites (RI = difference in beta value between 5th and 95th percentiles)."}
hachiyaPlt
```

```{r supp-qq-plots, dev="png", dpi=300, fig.asp=1.5, fig.cap="QQ-plots for EWAS results, including the full set of CpG sites passing quality control steps."}
par(mfrow=c(2,1), mar=c(3,2,2,1), mgp=c(2,0.5,0.5))
make_qqplot(whiResFull$p, plotTitle=paste0("WHI: lambda = ", gControl(whiResFull$p)))  # Q-Q & g. control
make_qqplot(fhsResFull$p, plotTitle=paste0("FHS: lambda = ", gControl(fhsResFull$p)))  # Q-Q & g. control
```

```{r supp-manhattan, warning=F, fig.cap="Manhattan plot of EWAS results in WHI."}
qqman::manhattan(qqmanInput_whi, suggestiveline=F,
                 genomewideline=-log10(0.05/nrow(whiRes)))
```

```{r supp-dmp-boxplot, fig.cap="Distribution of beta values in subjects who did and did not experience an incident CVD event."}
dmpBoxplot
```

```{r supp-genomic-enrichment, fig.asp=1.5, fig.cap="Relative proportions of CpG sites found in (a) gene-based, and (b) CpG island-based genomic regions. The background consists of the set of variable sites examined."}
plot_grid(geneGroupEnrichmentPlot, cpgIslandEnrichmentPlot, nrow=2, labels="auto")
```

```{r supp-gsea-table}
# gsaTopResTblAbbrevs <- mutate(gsaTopResTbl, 
#                               Term=ifelse(nchar(Term)<=40, Term, paste0(substr(Term,1,40),"...")))
kable(gsaTopResTbl, row.names=F, caption="GSEA results for CpGs with FDR<0.25 in either dataset.", 
      booktabs=T, longtable=T)
```

```{r supp-dmrs-whi}
kable(whiDmrs, caption="Top DMRs from WHI only.", booktabs=T, longtable=T)
```

```{r supp-dmrs-fhs}
kable(fhsDmrs, caption="Top DMRs from FHS only.", booktabs=T, longtable=T)
```

```{r supp-module-enrichment}
kable(allModuleEnrichment, row.names=F, caption="GO enrichment of CVD-related modules from WGCNA.",
      booktabs=T, longtable=T)
```

```{r supp-figs-and-tables, warning=F, dev="png", dpi=300, eval=F}
whiVarPlt
fhsVarPlt

hachiyaPlt  # Compare RIs of Hachiya vs. non-Hachiya sites in my cohorts

make_qqplot(whiResFull$p, plotTitle=paste0("WHI: lambda = ", gControl(whiResFull$p)))  # Q-Q & g. control
make_qqplot(fhsResFull$p, plotTitle=paste0("FHS: lambda = ", gControl(fhsResFull$p)))  # Q-Q & g. control

qqman::manhattan(qqmanInput_whi, suggestiveline=F,
                 genomewideline=-log10(0.05/nrow(whiRes)), main="WHI EWAS Results")

dmpBoxplot

geneGroupEnrichmentPlot
cpgIslandEnrichmentPlot

kable(gsaTopResTbl, caption="GSEA results for CpGs with FDR<0.25 in either dataset.", booktabs=T)

kable(whiDmrs, caption="Top DMRs from WHI only.", booktabs=T)
kable(fhsDmrs, caption="Top DMRs from FHS only.", booktabs=T)

kable(allModuleEnrichment, booktabs=T)
```

# References


```{r enrichment-hypotheses, eval=F}
vanBaakDataset <- readxl::read_excel("../int/13059_2017_1374_MOESM2_ESM.xlsx", sheet=2)
ess <- vanBaakDataset$CG[vanBaakDataset$`ESS hit`]
siv <- vanBaakDataset$CG[vanBaakDataset$`SIV hit`]

hitsInESS <- length(intersect(topCpGs, ess))
ess_depletion <- phyper(hitsInESS,  # Size of intersection
                        length(ess),  # Size of the CpG set of interest
                        nrow(betas_whi)-length(ess),  # Size of the complement of the set of interest
                        length(topCpGs),  # Number of hits being tested
                        lower.tail=T) 
print("ESS depletion p-value:")
print(ess_depletion)

hitsInSIV <- length(intersect(topCpGs, siv))
siv_depletion <- phyper(hitsInSIV,  # Size of intersection
                        length(siv),  # Size of the CpG set of interest
                        nrow(betas_whi)-length(siv),  # Size of the complement of the set of interest
                        length(topCpGs),  # Number of hits being tested
                        lower.tail=T)
print("SIV depletion p-value:")
print(siv_depletion)
```

```{r viz-sushi, eval=F}
resAnnoRegionWhi <- filter(resAnnoWhi, chr=="chr2", start>=chromstart, end<=chromend)
resAnnoRegionFhs <- filter(resAnnoFhs, chr=="chr2", start>=chromstart, end<=chromend)
plotDF <- rbind(resAnnoRegionWhi, resAnnoRegionFhs) %>%
  mutate()
plotBed(rbind(resAnnoRegionWhi, resAnnoRegionFhs), 
        chrom=chrom, chromstart=chromstart, chromend=chromend, type="circles",
        color=rep(c("red","blue"),times=c(nrow(resAnnoRegionWhi),nrow(resAnnoRegionFhs))), pch="*")
legend("topright", legend=c("WHI","FHS"), fill=c("red","blue"))
zoomsregion(dmr, highlight=T)
labelgenome(chrom, chromstart, chromend, n=5, scale="Kb")
mtext("-log10(p-value)", side=2, line=2)
axis(side=2)
```

```{r train-test-funcs, eval=F}
train_mrs <- function(mod, meta, meth, trainset, 
                      alpha=0.5, nfolds=5, lmr=0.1, lambda.type="lambda.min", parallel=F) {
  covars_frame <- model.frame(as.formula(mod), meta, na.action=na.pass)  # To allow NAs to pass through
  covars_mat <- model.matrix(as.formula(mod), covars_frame)[,-1]  # Converts factors to numeric
  design_mat <- cbind(covars_mat, t(meth))
  cvd_surv <- Surv(time=meta$time, event=meta$event)  # Outcome
  mrs.fit.cv <- cv.glmnet(design_mat[trainset,], cvd_surv[trainset], family="cox", alpha=alpha,  # Train MRS model
                          penalty.factor=ifelse(grepl("^cg", colnames(design_mat)), 1, 0), 
                          nfolds=nfolds, lambda.min.ratio=lmr, parallel=parallel)
  coefs <- coef(mrs.fit.cv, s=lambda.type)
  cpgCoefs <- coefs[grepl("^cg", rownames(coefs)),]
  list(fit=mrs.fit.cv,  # Output is a list containing the model fit and the CpG model coefficients
       allCoefs=coefs,
       cpgCoefs=cpgCoefs)
}

calc_mrs <- function(coefs, meth) {
  if (length(coefs)==0 || all(coefs==0)) {
    message("No non-zero coefficients")
    return(rep(NA, ncol(meth)))
  } else {
    nonZeroCoefs <- coefs[coefs!=0]
    as.vector(t(meth[names(nonZeroCoefs),,drop=F]) %*% nonZeroCoefs)
  }
} 

test_mrs <- function(meta, mrs, testset) {
  meta$mrs <- scale(mrs)
  meta$cvd_surv <- Surv(time=meta$time, event=meta$event)  # Outcome
  mrs.test <- coxph(cvd_surv~mrs, data=meta, subset=testset) # Test the MRS using a Cox model
  mrs.res <- summary(mrs.test)$coef["mrs",c("exp(coef)","z")]
  tibble(HR_per_SD=mrs.res["exp(coef)"], p=as.character(2*pnorm(-abs(mrs.res["z"]))))
}

trainTestMRS <- function(mod, meta, meth, trainset, testset, alpha=0.5, nfolds=5, lmr=0.1, lt="lambda.min", parallel=F) {
  mrs_fit <- train_mrs(mod, meta, meth, trainset, alpha=alpha, nfolds=nfolds, lmr=lmr, lambda.type=lt, parallel=parallel)
  mrs_values <- calc_mrs(mrs_fit$cpgCoefs, meth)
  mrs_test <- tryCatch(test_mrs(meta, mrs_values, testset), error=function(e) matrix(NA, 1, 2))
  list(fit=mrs_fit, components=mrs_fit$cpgCoefs, values=mrs_values, res=mrs_test)
}
```

```{r bayesian-network, message=F, eval=F}
library(bnlearn)

sbp_cpgs <- c(PHGDH="cg14476101", TXNIP="cg19693031")
hdl_cpgs <- c(DHCR24="cg17901584", ABCG1="cg06500161")
tg_cpgs <- c(CPT1A_1="cg00574958", CPT1A_2="cg17058475")
tc_cpgs <- c(SQLE="cg00285394", NLRC5="cg07839457")

sbp_cpgs <- c(PHGDH="cg14476101")
hdl_cpgs <- c(ABCG1="cg06500161")
tg_cpgs <- c(CPT1A_1="cg00574958")
tc_cpgs <- c(SQLE="cg00285394")
acs_cpgs <- c(RUNX3="cg21217270", DGKQ="cg12226453", RGS3="cg14049634", ANKRD11="cg27513667")
cvdSR_cpgs <- c(GRIP1="cg09414535", KCNJ14="cg10222534", PKD2="cg26215428", AIM2="cg10636246",
                HRH2="cg14345676", NGEF="cg19485804", TNS1="cg18752854")

rfs <- c("sbp","chol","hdl","tg")

bnInputData <- cbind(select(nmd_fhs, age, sex, bmi, one_of(rfs), event, pastEvent), 
           t(betas_fhs[unique(c(topCpgs, sbp_cpgs, hdl_cpgs, tg_cpgs, tc_cpgs)),])) %>%
  na.omit() %>%
  mutate_at(vars(one_of(rfs)), as.numeric) %>%
  mutate(sex=factor(sex), event=factor(event), pastEvent=factor(pastEvent), age=as.numeric(age))

ewasCpgs <- fhsRes$CpG[fhsRes$CpG %in% topCpgs][1:5]
ofInterest <- c(age="age", bmi="bmi", pastEvent="pastEvent", rfs, sbp_cpgs, hdl_cpgs, tg_cpgs, tc_cpgs)

bn <- bnlearn::rsmax2(select(bnInputData, one_of("age", "bmi", "pastEvent",
                                                 # sbp_cpgs, hdl_cpgs, tg_cpgs, tc_cpgs,
                                                 ewasCpgs,
                                                 # topCpgs[grepl("dmr",names(topCpgs))],
                                                 c("chol","hdl","tg","sbp"))))
bnPlt <- plot(bn, 
              highlight=ewasCpgs,
              # highlight=topCpgs[grepl("dmr",names(topCpgs))],
              main="Bayesian network of our CpGs, RF-related CpGs, and RFs")
bnPlt
```

```{r weighted-sum-score, eval=F}
whiFdr <- na.omit(whiRes[whiRes$fdr<0.05,])
whiEwasCoefs <- setNames(whiFdr$coef, whiFdr$CpG)
ewasMrs <- calc_mrs(coefs=whiEwasCoefs, meth=betas_fhs)

# Basic evaluation
binFit <- glm(event~ewasMrs, data=nmd_fhs, family="binomial")
summary(binFit)

roc(nmd_fhs$event~ewasMrs)

# Comparison to FRS
source("helpers.R")
nmd_fhs <- nmd_fhs %>%
  mutate(diabetes=dm_med|glu>125) %>%
  replace_na(list(diabetes=F))
frs <- calc_FRS(nmd_fhs)
frsModel <- glm(event~frs, data=nmd_fhs)
frsPlusMrsModel <- glm(event~frs+ewasMrs, data=nmd_fhs)
library(nricens)
nriCalc <- nribin(mdl.std=frsModel, mdl.new=frsPlusMrsModel)$nri


```

<!--
* Proportional hazards checks for each individual regression? Only for replication cohort?
* Relate to existing findings for prevalent CVD
* Differential variability
-->


